{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76cde750-8ccd-434f-82f8-94e84d6a6b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "using BenchmarkTools\n",
    "using Revise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "983f0dc7-21c5-4531-a5c2-6790849cf78d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling FeatureSetEnrichment [d7bef1b9-4643-4ed1-9ceb-ed43dd382ec0]\n",
      "└ @ Base loading.jl:1317\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<script>\n",
       "// Immediately-invoked-function-expression to avoid global variables.\n",
       "(function() {\n",
       "    var warning_div = document.getElementById(\"webio-warning-994060335855289353\");\n",
       "    var hide = function () {\n",
       "        var script = document.getElementById(\"webio-setup-1617827721267030514\");\n",
       "        var parent = script && script.parentElement;\n",
       "        var grandparent = parent && parent.parentElement;\n",
       "        if (grandparent) {\n",
       "            grandparent.style.display = \"none\";\n",
       "        }\n",
       "        warning_div.style.display = \"none\";\n",
       "    };\n",
       "    if (typeof Jupyter !== \"undefined\") {\n",
       "        console.log(\"WebIO detected Jupyter notebook environment.\");\n",
       "        // Jupyter notebook.\n",
       "        var extensions = (\n",
       "            Jupyter\n",
       "            && Jupyter.notebook.config.data\n",
       "            && Jupyter.notebook.config.data.load_extensions\n",
       "        );\n",
       "        if (extensions && extensions[\"webio-jupyter-notebook\"]) {\n",
       "            // Extension already loaded.\n",
       "            console.log(\"Jupyter WebIO nbextension detected; not loading ad-hoc.\");\n",
       "            hide();\n",
       "            return;\n",
       "        }\n",
       "    } else if (window.location.pathname.includes(\"/lab\")) {\n",
       "        // Guessing JupyterLa\n",
       "        console.log(\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\");\n",
       "        hide();\n",
       "        return;\n",
       "    }\n",
       "})();\n",
       "\n",
       "</script>\n",
       "<p\n",
       "    id=\"webio-warning-994060335855289353\"\n",
       "    class=\"output_text output_stderr\"\n",
       "    style=\"padding: 1em; font-weight: bold;\"\n",
       ">\n",
       "    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\n",
       "    For troubleshooting, please see <a href=\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\">\n",
       "    the WebIO/IJulia documentation</a>.\n",
       "    <!-- TODO: link to installation docs. -->\n",
       "</p>\n"
      ],
      "text/plain": [
       "HTML{String}(\"<script>\\n// Immediately-invoked-function-expression to avoid global variables.\\n(function() {\\n    var warning_div = document.getElementById(\\\"webio-warning-994060335855289353\\\");\\n    var hide = function () {\\n        var script = document.getElementById(\\\"webio-setup-1617827721267030514\\\");\\n        var parent = script && script.parentElement;\\n        var grandparent = parent && parent.parentElement;\\n        if (grandparent) {\\n            grandparent.style.display = \\\"none\\\";\\n        }\\n        warning_div.style.display = \\\"none\\\";\\n    };\\n    if (typeof Jupyter !== \\\"undefined\\\") {\\n        console.log(\\\"WebIO detected Jupyter notebook environment.\\\");\\n        // Jupyter notebook.\\n        var extensions = (\\n            Jupyter\\n            && Jupyter.notebook.config.data\\n            && Jupyter.notebook.config.data.load_extensions\\n        );\\n        if (extensions && extensions[\\\"webio-jupyter-notebook\\\"]) {\\n            // Extension already loaded.\\n            console.log(\\\"Jupyter WebIO nbextension detected; not loading ad-hoc.\\\");\\n            hide();\\n            return;\\n        }\\n    } else if (window.location.pathname.includes(\\\"/lab\\\")) {\\n        // Guessing JupyterLa\\n        console.log(\\\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\\\");\\n        hide();\\n        return;\\n    }\\n})();\\n\\n</script>\\n<p\\n    id=\\\"webio-warning-994060335855289353\\\"\\n    class=\\\"output_text output_stderr\\\"\\n    style=\\\"padding: 1em; font-weight: bold;\\\"\\n>\\n    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\\n    For troubleshooting, please see <a href=\\\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\\\">\\n    the WebIO/IJulia documentation</a>.\\n    <!-- TODO: link to installation docs. -->\\n</p>\\n\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "using CSV\n",
    "using DataFrames\n",
    "using JSON\n",
    "\n",
    "using DataIO\n",
    "using GCTGMT\n",
    "using Normalization\n",
    "using Support\n",
    "\n",
    "using FeatureSetEnrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f25618fe-79dc-49e0-bee4-6ac273975ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "name_to_set_to_element_ = Dict(\n",
    "    splitpath(path)[end] => read_gmt(path) for\n",
    "    path in read_directory(\"/Users/kwat/Downloads/gene_sets/\")\n",
    ")\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24b75e0a-3c53-4fd9-aa33-71edd712f6db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n"
     ]
    }
   ],
   "source": [
    "method_ = Vector{String}()\n",
    "\n",
    "for method in keys(score_set_new([\"a\", \"b\"], [-1.0, 1.0], [\"a\"]; plot = false))\n",
    "\n",
    "    push!(method_, string(method, \" extreme\"), string(method, \" area\"))\n",
    "\n",
    "end\n",
    "\n",
    "n_method = length(method_)\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "028f61ac-6359-49db-94c4-cfd07db70a26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/Users/kwat/Downloads/benchmarks/\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmarks_directory_path = \"/Users/kwat/Downloads/benchmarks/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25f54201-7b3d-4edf-93cd-ce9b5415b07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_ = Vector{String}()\n",
    "\n",
    "benchmark_size_ = Vector{Int64}()\n",
    "\n",
    "benchmark_x_method_rows = Vector{Vector{Float64}}()\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc898b0f-e4bc-4a4e-ba7c-facd6d37730f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "    4951 set.\n",
      "    AIGNER_ZEB1_TARGETS (35)\n",
      "    SANSOM_APC_TARGETS_UP (126)\n",
      "    BCAT_GDS748_UP (48)\n",
      "    BCAT.100_UP.V1_UP (49)\n",
      "    PID_WNT_SIGNALING_PATHWAY (28)\n",
      "    LIU_CDX2_TARGETS_UP (36)\n",
      "    LUCAS_HNF4A_TARGETS_UP (58)\n",
      "    BCAT_GDS748 (missing)\n",
      "10\n",
      "    4762 set.\n",
      "    LANG_MYB_FAMILY_TARGETS (29)\n",
      "    PID_CMYB_PATHWAY (84)\n",
      "100\n",
      "    10679 set.\n",
      "    GO_CELL_CYCLE_G1_S_PHASE_TRANSITION (111)\n",
      "    REACTOME_M_G1_TRANSITION (81)\n",
      "    REACTOME_MITOTIC_M_M_G1_PHASES (172)\n",
      "    KEGG_DNA_REPLICATION (36)\n",
      "    REACTOME_DNA_REPLICATION (192)\n",
      "    EGUCHI_CELL_CYCLE_RB1_TARGETS (23)\n",
      "    WHITFIELD_CELL_CYCLE_LITERATURE (44)\n",
      "101\n",
      "    4762 set.\n",
      "    BIOCARTA_NFKB_PATHWAY (23)\n",
      "    TIAN_TNF_SIGNALING_VIA_NFKB (28)\n",
      "    BIOCARTA_TNFR2_PATHWAY (18)\n",
      "    SANA_TNF_SIGNALING_UP (83)\n",
      "    PHONG_TNF_TARGETS_UP (63)\n",
      "    SCHOEN_NFKB_SIGNALING (34)\n",
      "    GILMORE_CORE_NFKB_PATHWAY (14)\n",
      "102\n",
      "    4762 set.\n",
      "    REACTOME_ALPHA_LINOLENIC_ACID_ALA_METABOLISM (12)\n",
      "    KEGG_BETA_ALANINE_METABOLISM (22)\n",
      "103\n",
      "    4762 set.\n",
      "    REACTOME_CELL_CYCLE (421)\n",
      "    EGUCHI_CELL_CYCLE_RB1_TARGETS (23)\n",
      "    REACTOME_REGULATION_OF_MITOTIC_CELL_CYCLE (85)\n",
      "    RAMALHO_STEMNESS (missing)\n",
      "104\n",
      "    4811 set.\n",
      "    ESW_FLI_UPREGULATED (572)\n",
      "    TORCHIA_TARGETS_OF_EWSR1_FLI1_FUSION_TOP20 (missing)\n",
      "    SILIGAN_TARGETS_OF_EWS_FLI1_FUSION (missing)\n",
      "105\n",
      "    6229 set.\n",
      "    REACTOME_THE_CITRIC_ACID_TCA_CYCLE_AND_RESPIRATORY_ELECTRON_TRANSPORT (175)\n",
      "    MOOTHA_VOXPHOS (87)\n",
      "    KEGG_OXIDATIVE_PHOSPHORYLATION (132)\n",
      "106\n",
      "    4812 set.\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "    MOSERLE_IFNA_RESPONSE (31)\n",
      "    SANA_RESPONSE_TO_IFNG (missing)\n",
      "    GRANDVAUX_IFN_RESPONSE_NOT_VIA_IRF3 (14)\n",
      "    REACTOME_RIG_I_MDA5_MEDIATED_INDUCTION_OF_IFN_ALPHA_BETA_PATHWAYS (73)\n",
      "    HECKER_IFNB1_TARGETS (95)\n",
      "107\n",
      "    4872 set.\n",
      "    GSE7852_LN_VS_FAT_TCONV_DN (200)\n",
      "    GSE7852_LN_VS_FAT_TREG_DN (200)\n",
      "    GSE3720_UNSTIM_VS_LPS_STIM_VD2_GAMMADELTA_TCELL_DN (148)\n",
      "    GSE14769_UNSTIM_VS_60MIN_LPS_BMDM_DN (200)\n",
      "    GSE9988_LPS_VS_VEHICLE_TREATED_MONOCYTE_UP (200)\n",
      "    GSE9988_ANTI_TREM1_AND_LPS_VS_VEHICLE_TREATED_MONOCYTES (missing)\n",
      "    GSE9988_LPS_VS_CTRL_TREATED_MONOCYTE (missing)\n",
      "    GSE9988_LPS_VS_CTRL_TREATED_MONOCYTE_UP (200)\n",
      "108\n",
      "    6229 set.\n",
      "    CAIRO_HEPATOBLASTOMA_UP (213)\n",
      "    CAIRO_HEPATOBLASTOMA_DN (267)\n",
      "    CAIRO_HEPATOBLASTOMA_CLASSES_UP (612)\n",
      "    CAIRO_LIVER_DEVELOPMENT_UP (166)\n",
      "    HSIAO_LIVER_SPECIFIC_GENES (250)\n",
      "109\n",
      "    6229 set.\n",
      "    WONG_EMBRYONIC_STEM_CELL_CORE (335)\n",
      "    BHATTACHARYA_EMBRYONIC_STEM_CELL (84)\n",
      "11\n",
      "    4762 set.\n",
      "    XU_AKT1_TARGETS_48HR (9)\n",
      "110\n",
      "    14815 set.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "    HALLMARK_ALLOGRAFT_REJECTION (200)\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    GO_INTERFERON_GAMMA_MEDIATED_SIGNALING_PATHWAY (91)\n",
      "    GO_CELLULAR_RESPONSE_TO_INTERFERON_BETA (22)\n",
      "    GO_POSITIVE_REGULATION_OF_INTERFERON_ALPHA_PRODUCTION (20)\n",
      "111\n",
      "    4762 set.\n",
      "    REACTOME_TRANSPORT_OF_RIBONUCLEOPROTEINS_INTO_THE_HOST_NUCLEUS (27)\n",
      "    REACTOME_PYRIMIDINE_METABOLISM (24)\n",
      "    KEGG_ONE_CARBON_POOL_BY_FOLATE (17)\n",
      "112\n",
      "    6229 set.\n",
      "    WONG_MITOCHONDRIA_GENE_MODULE (218)\n",
      "    MOOTHA_VOXPHOS (87)\n",
      "    MOOTHA_HUMAN_MITODB_6_2002 (431)\n",
      "    REACTOME_MITOCHONDRIAL_PROTEIN_IMPORT (65)\n",
      "    STEIN_ESRRA_TARGETS_UP (385)\n",
      "    KEGG_RIBOSOME (88)\n",
      "    REACTOME_TRANSLATION (295)\n",
      "113\n",
      "    9684 set.\n",
      "    AMIT_DELAYED_EARLY_GENES (18)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "    NAGASHIMA_NRG1_SIGNALING_UP (176)\n",
      "    TIAN_TNF_SIGNALING_NOT_VIA_NFKB (22)\n",
      "    RASHI_NFKB1_TARGETS (19)\n",
      "    DUTTA_APOPTOSIS_VIA_NFKB (33)\n",
      "    PHONG_TNF_TARGETS_UP (63)\n",
      "114\n",
      "    10679 set.\n",
      "    GO_T_CELL_ACTIVATION_INVOLVED_IN_IMMUNE_RESPONSE (60)\n",
      "    GO_ALPHA_BETA_T_CELL_ACTIVATION (54)\n",
      "    GO_IMMUNE_RESPONSE (1100)\n",
      "    GO_CELLULAR_RESPONSE_TO_INTERFERON_GAMMA (122)\n",
      "    GO_RESPONSE_TO_TYPE_I_INTERFERON (68)\n",
      "    GO_REGULATION_OF_INTERFERON_ALPHA_PRODUCTION (19)\n",
      "    SANA_RESPONSE_TO_IFNG (missing)\n",
      "    GRANDVAUX_IFN_RESPONSE_NOT_VIA_IRF3 (14)\n",
      "    MOSERLE_IFNA_RESPONSE (31)\n",
      "115\n",
      "    6229 set.\n",
      "    ZHAN_MULTIPLE_MYELOMA_PR_UP (46)\n",
      "    EGUCHI_CELL_CYCLE_RB1_TARGETS (22)\n",
      "    ZHOU_CELL_CYCLE_GENES_IN_IR_RESPONSE_6HR (84)\n",
      "    BENPORATH_PROLIFERATION (147)\n",
      "116\n",
      "    5787 set.\n",
      "    BACH1_01 (263)\n",
      "    NRF2_Q4 (255)\n",
      "    SINGH_NFE2L2_TARGETS (15)\n",
      "    NFE2L2.V2KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS (missing)\n",
      "    KEGG_PENTOSE_PHOSPHATE_PATHWAY (27)\n",
      "117\n",
      "    10729 set.\n",
      "    REACTOME_CHOLESTEROL_BIOSYNTHESIS (24)\n",
      "    HALLMARK_CHOLESTEROL_HOMEOSTASIS (74)\n",
      "    KEGG_STEROID_BIOSYNTHESIS (17)\n",
      "    GO_STEROID_BIOSYNTHETIC_PROCESS (114)\n",
      "118\n",
      "    5529 set.\n",
      "    HORTON_SREBF_TARGETS (24)\n",
      "    REACTOME_ACTIVATION_OF_GENE_EXPRESSION_BY_SREBF_SREBP (42)\n",
      "119\n",
      "    6278 set.\n",
      "    alveolar_KRAS_up (572)\n",
      "    JI_CARCINOGENESIS_BY_KRAS_AND_STK11_UP (12)\n",
      "12\n",
      "    189 set.\n",
      "    GLI1_UP.V1_UP (27)\n",
      "120\n",
      "    14765 set.\n",
      "    GO_NEGATIVE_REGULATION_OF_VIRAL_ENTRY_INTO_HOST_CELL (18)\n",
      "    GO_UDP_N_ACETYLGLUCOSAMINE_METABOLIC_PROCESS (16)\n",
      "    GO_NATURAL_KILLER_CELL_MEDIATED_IMMUNITY (66)\n",
      "121\n",
      "    6468 set.\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE (57)\n",
      "    PDGF_UP.V1_DN (135)\n",
      "    NRL_DN.V1_UP (136)\n",
      "    KRAS.300_UP.V1_UP (142)\n",
      "    MTOR_UP.N4.V1_UP (196)\n",
      "    IL21_UP.V1_UP (185)\n",
      "    EGFR_UP.V1_UP (192)\n",
      "    IL15_UP.V1_UP (189)\n",
      "    PTEN_DN.V2_UP (139)\n",
      "    MTOR_UP.V1_DN (181)\n",
      "    E2F3_UP.V1_UP (191)\n",
      "    JAK2_DN.V1_DN (145)\n",
      "    BMI1_DN_MEL18_DN.V1_UP (145)\n",
      "    P53_DN.V2_UP (148)\n",
      "    BRCA1_DN.V1_DN (139)\n",
      "    CAHOY_ASTROGLIAL (100)\n",
      "    KRAS.50_UP.V1_UP (48)\n",
      "    TGFB_UP.V1_DN (191)\n",
      "    CSR_LATE_UP.V1_DN (147)\n",
      "122\n",
      "    5529 set.\n",
      "    HINATA_NFKB_TARGETS_KERATINOCYTE_UP (85)\n",
      "    TIAN_TNF_SIGNALING_VIA_NFKB (26)\n",
      "    REACTOME_SENESCENCE_ASSOCIATED_SECRETORY_PHENOTYPE_SASP (111)\n",
      "    RASHI_NFKB1_TARGETS (17)\n",
      "    GILMORE_CORE_NFKB_PATHWAY (12)\n",
      "    BIOCARTA_NFKB_PATHWAY (21)\n",
      "    KAMMINGA_SENESCENCE (40)\n",
      "123\n",
      "    6279 set.\n",
      "    HALLMARK_TGF_BETA_SIGNALING (54)\n",
      "    BIOCARTA_TGFB_PATHWAY (19)\n",
      "    REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX_IN_CANCER (8)\n",
      "    WP_CANONICAL_AND_NONCANONICAL_TGFB_SIGNALING (17)\n",
      "124\n",
      "    21044 set.\n",
      "    BIOCARTA_IL1R_PATHWAY (31)\n",
      "    REACTOME_COMPLEMENT_CASCADE (115)\n",
      "    HALLMARK_IL6_JAK_STAT3_SIGNALING (87)\n",
      "    GO_T_HELPER_1_CELL_CYTOKINE_PRODUCTION (8)\n",
      "    ST_JAK_STAT_PATHWAY (9)\n",
      "125\n",
      "    4762 set.\n",
      "    REACTOME_TRAF6_MEDIATED_IRF7_ACTIVATION_IN_TLR7_8_OR_9_SIGNALING (10)\n",
      "    REACTOME_TRAFFICKING_AND_PROCESSING_OF_ENDOSOMAL_TLR (14)\n",
      "126\n",
      "    9785 set.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (71)\n",
      "    REACTOME_INTERFERON_GAMMA_SIGNALING (93)\n",
      "    EGR2_01 (198)\n",
      "    REACTOME_EGR2_AND_SOX10_MEDIATED_INITIATION_OF_SCHWANN_CELL_MYELINATION (29)\n",
      "127\n",
      "    6418 set.\n",
      "    BCAT_BILD_ET_AL_UP (45)\n",
      "128\n",
      "    6418 set.\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE (57)\n",
      "    REACTOME_SIGNALING_BY_HIPPO (20)\n",
      "129\n",
      "    6279 set.\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    COLLER_MYC_TARGETS_UP (24)\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    KALMA_E2F1_TARGETS (11)\n",
      "13\n",
      "    4762 set.\n",
      "    ODONNELL_TARGETS_OF_MYC_AND_TFRC_UP (83)\n",
      "130\n",
      "    6279 set.\n",
      "    HALLMARK_MTORC1_SIGNALING (200)\n",
      "    HALLMARK_OXIDATIVE_PHOSPHORYLATION (200)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "131\n",
      "    4762 set.\n",
      "    KEGG_RIBOSOME (88)\n",
      "    KEGG_SPLICEOSOME (128)\n",
      "132\n",
      "    4762 set.\n",
      "    KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION (267)\n",
      "    BIOCARTA_CYTOKINE_PATHWAY (22)\n",
      "    KEGG_CHEMOKINE_SIGNALING_PATHWAY (190)\n",
      "    HALLMARK_IL6_JAK_STAT3_SIGNALING (missing)\n",
      "133\n",
      "    6278 set.\n",
      "    Bramswig_Alpha_Strong.genes (572)\n",
      "    Muraro_ALpha.genes (572)\n",
      "    Wang_adult_alpha.genes (572)\n",
      "134\n",
      "    4762 set.\n",
      "    REACTOME_P53_INDEPENDENT_G1_S_DNA_DAMAGE_CHECKPOINT (51)\n",
      "    RASHI_RESPONSE_TO_IONIZING_RADIATION_2 (127)\n",
      "135\n",
      "    4812 set.\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "136\n",
      "    5917 set.\n",
      "    GO_MITOTIC_NUCLEAR_DIVISION (361)\n",
      "    GO_MITOTIC_RECOMBINATION (41)\n",
      "    GO_MITOTIC_SPINDLE_ORGANIZATION (69)\n",
      "    GO_SISTER_CHROMATID_COHESION (111)\n",
      "    GO_CELL_CYCLE_PHASE_TRANSITION (255)\n",
      "137\n",
      "    6229 set.\n",
      "    BIOCARTA_PTEN_PATHWAY (18)\n",
      "    SA_PTEN_PATHWAY (17)\n",
      "138\n",
      "    10679 set.\n",
      "    GO_ORGANIC_ACID_METABOLIC_PROCESS (953)\n",
      "    GO_MONOCARBOXYLIC_ACID_METABOLIC_PROCESS (503)\n",
      "    PID_HNF3B_PATHWAY (45)\n",
      "139\n",
      "    4812 set.\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    SARRIO_EPITHELIAL_MESENCHYMAL_TRANSITION_UP (180)\n",
      "14\n",
      "    4762 set.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "    PID_NFKAPPAB_ATYPICAL_PATHWAY (17)\n",
      "140\n",
      "    6172 set.\n",
      "    IPA_PAX5_UP (817)\n",
      "    GSE9878_CTRL_VS_EBF_TRANSDUCED_PAX5_KO_PRO_BCELL_DN (200)\n",
      "141\n",
      "    6829 set.\n",
      "    REACTOME_EGR2_AND_SOX10_MEDIATED_INITIATION_OF_SCHWANN_CELL_MYELINATION (29)\n",
      "    IPA_SOX10_UP (817)\n",
      "142\n",
      "    6829 set.\n",
      "    LUCAS_HNF4A_TARGETS_UP (58)\n",
      "    IPA_HNF4A_UP (817)\n",
      "    SUMI_HNF4A_TARGETS (29)\n",
      "    OHGUCHI_LIVER_HNF4A_TARGETS_DN (167)\n",
      "143\n",
      "    6829 set.\n",
      "    IPA_SPI1_UP (817)\n",
      "144\n",
      "    6829 set.\n",
      "    IPA_MEIS1_UP (817)\n",
      "145\n",
      "    5529 set.\n",
      "    ONDER_CDH1_TARGETS_2_DN (473)\n",
      "    ONDER_CDH1_TARGETS_3_DN (59)\n",
      "    ONDER_CDH1_TARGETS_1_DN (170)\n",
      "146\n",
      "    10192 set.\n",
      "    GO_MPP7_DLG1_LIN7_COMPLEX (5)\n",
      "147\n",
      "    15721 set.\n",
      "    REACTOME_GRB7_EVENTS_IN_ERBB2_SIGNALING (5)\n",
      "    REACTOME_ERBB2_ACTIVATES_PTK6_SIGNALING (13)\n",
      "    GO_ERBB2_SIGNALING_PATHWAY (32)\n",
      "    LANDIS_ERBB2_BREAST_TUMORS_324_UP (151)\n",
      "    REACTOME_PI3K_EVENTS_IN_ERBB2_SIGNALING (16)\n",
      "148\n",
      "    6829 set.\n",
      "    IPA_FOXA1_UP (817)\n",
      "149\n",
      "    5529 set.\n",
      "    REACTOME_RUNX3_REGULATES_BCL2L11_BIM_TRANSCRIPTION (5)\n",
      "    REACTOME_RUNX3_REGULATES_IMMUNE_RESPONSE_AND_CELL_MIGRATION (6)\n",
      "15\n",
      "    4762 set.\n",
      "    PID_IL2_STAT5_PATHWAY (30)\n",
      "    SCHURINGA_STAT5A_TARGETS_UP (21)\n",
      "    FUNG_IL2_TARGETS_WITH_STAT5_BINDING_SITES (7)\n",
      "150\n",
      "    5529 set.\n",
      "    REACTOME_SHC1_EVENTS_IN_EGFR_SIGNALING (14)\n",
      "    REACTOME_EGFR_DOWNREGULATION (31)\n",
      "    REACTOME_EGFR_INTERACTS_WITH_PHOSPHOLIPASE_C_GAMMA (9)\n",
      "    REACTOME_SIGNALING_BY_EGFR_IN_CANCER (25)\n",
      "    REACTOME_EGFR_TRANSACTIVATION_BY_GASTRIN (9)\n",
      "151\n",
      "    5529 set.\n",
      "    REACTOME_MET_INTERACTS_WITH_TNS_PROTEINS (5)\n",
      "    REACTOME_MET_ACTIVATES_PI3K_AKT_SIGNALING (6)\n",
      "    REACTOME_MET_PROMOTES_CELL_MOTILITY (41)\n",
      "    REACTOME_MET_ACTIVATES_PTK2_SIGNALING (30)\n",
      "152\n",
      "    5529 set.\n",
      "    REACTOME_DEFECTIVE_GALNT3_CAUSES_FAMILIAL_HYPERPHOSPHATEMIC_TUMORAL_CALCINOSIS_HFTC (16)\n",
      "153\n",
      "    6829 set.\n",
      "    IPA_MECOM_UP (817)\n",
      "154\n",
      "    5529 set.\n",
      "    REACTOME_NEGATIVE_REGULATORS_OF_DDX58_IFIH1_SIGNALING (34)\n",
      "155\n",
      "    6829 set.\n",
      "    REACTOME_CD28_DEPENDENT_VAV1_PATHWAY (12)\n",
      "156\n",
      "    5529 set.\n",
      "    REACTOME_RUNX3_REGULATES_YAP1_MEDIATED_TRANSCRIPTION (8)\n",
      "    REACTOME_YAP1_AND_WWTR1_TAZ_STIMULATED_GENE_EXPRESSION (15)\n",
      "157\n",
      "    5529 set.\n",
      "    REACTOME_PCNA_DEPENDENT_LONG_PATCH_BASE_EXCISION_REPAIR (21)\n",
      "158\n",
      "    5529 set.\n",
      "    KOYAMA_SEMA3B_TARGETS_UP (264)\n",
      "159\n",
      "    6829 set.\n",
      "    SHAFFER_IRF4_TARGETS_IN_ACTIVATED_DENDRITIC_CELL (64)\n",
      "    SHAFFER_IRF4_MULTIPLE_MYELOMA_PROGRAM (36)\n",
      "    SHAFFER_IRF4_TARGETS_IN_MYELOMA_VS_MATURE_B_LYMPHOCYTE (100)\n",
      "    IPA_IRF4_UP (817)\n",
      "16\n",
      "    4812 set.\n",
      "    ST_STAT3_PATHWAY (11)\n",
      "    HALLMARK_IL6_JAK_STAT3_SIGNALING (87)\n",
      "    AZARE_STAT3_TARGETS (24)\n",
      "    DAUER_STAT3_TARGETS_UP (49)\n",
      "160\n",
      "    6829 set.\n",
      "    IPA_TP63_UP (817)\n",
      "161\n",
      "    11701 set.\n",
      "    GSE24972_WT_VS_IRF8_KO_SPLEEN_FOLLICULAR_BCELL_DN (200)\n",
      "    IPA_IRF8_UP (817)\n",
      "    IPA_IRF8_DN (817)\n",
      "162\n",
      "    6829 set.\n",
      "    IPA_CDX2_DN (817)\n",
      "    LIU_CDX2_TARGETS_UP (35)\n",
      "    IPA_CDX2_UP (817)\n",
      "163\n",
      "    5529 set.\n",
      "    REACTOME_ROLE_OF_LAT2_NTAL_LAB_ON_CALCIUM_MOBILIZATION (71)\n",
      "164\n",
      "    6829 set.\n",
      "    SINGH_NFE2L2_TARGETS (14)\n",
      "    IPA_NFE2L2_UP (817)\n",
      "165\n",
      "    5529 set.\n",
      "    REACTOME_ACTIVATION_OF_C3_AND_C5 (8)\n",
      "166\n",
      "    5529 set.\n",
      "    REACTOME_CD22_MEDIATED_BCR_REGULATION (61)\n",
      "167\n",
      "    5529 set.\n",
      "    SHI_SPARC_TARGETS_DN (13)\n",
      "168\n",
      "    5529 set.\n",
      "    REACTOME_YAP1_AND_WWTR1_TAZ_STIMULATED_GENE_EXPRESSION (15)\n",
      "169\n",
      "    6829 set.\n",
      "    IPA_FOSL1_UP (817)\n",
      "17\n",
      "    4762 set.\n",
      "    BIOCARTA_BARRESTIN_SRC_PATHWAY (15)\n",
      "    BILD_SRC_ONCOGENIC_SIGNATURE (62)\n",
      "170\n",
      "    6829 set.\n",
      "    IPA_CTNNB1_UP (817)\n",
      "    FEVR_CTNNB1_TARGETS_UP (686)\n",
      "171\n",
      "    5529 set.\n",
      "    REACTOME_GRB7_EVENTS_IN_ERBB2_SIGNALING (5)\n",
      "172\n",
      "    6829 set.\n",
      "    IPA_HNF1A_UP (817)\n",
      "    SERVITJA_ISLET_HNF1A_TARGETS_DN (107)\n",
      "173\n",
      "    5529 set.\n",
      "    ROZANOV_MMP14_TARGETS_UP (267)\n",
      "    ROZANOV_MMP14_TARGETS_SUBSET (33)\n",
      "174\n",
      "    5529 set.\n",
      "    REACTOME_RUNX1_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_BCR_SIGNALING (6)\n",
      "    REACTOME_RUNX1_REGULATES_ESTROGEN_RECEPTOR_MEDIATED_TRANSCRIPTION (6)\n",
      "    REACTOME_RUNX1_AND_FOXP3_CONTROL_THE_DEVELOPMENT_OF_REGULATORY_T_LYMPHOCYTES_TREGS (10)\n",
      "    REACTOME_RUNX1_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_DIFFERENTIATION_OF_MYELOID_CELLS (7)\n",
      "175\n",
      "    6829 set.\n",
      "    IPA_FOXA2_UP (817)\n",
      "176\n",
      "    5529 set.\n",
      "    REACTOME_DEFECTIVE_EXT2_CAUSES_EXOSTOSES_2 (14)\n",
      "177\n",
      "    10192 set.\n",
      "    GO_MPP7_DLG1_LIN7_COMPLEX (5)\n",
      "178\n",
      "    6829 set.\n",
      "    IPA_KLF5_UP (817)\n",
      "179\n",
      "    5529 set.\n",
      "    REACTOME_NRCAM_INTERACTIONS (7)\n",
      "18\n",
      "    4762 set.\n",
      "    KOINUMA_TARGETS_OF_SMAD2_OR_SMAD3 (824)\n",
      "    REACTOME_TRANSCRIPTIONAL_ACTIVITY_OF_SMAD2_SMAD3_SMAD4_HETEROTRIMER (38)\n",
      "180\n",
      "    6829 set.\n",
      "    IPA_NFIX_UP (817)\n",
      "181\n",
      "    6829 set.\n",
      "    IPA_POU2F2_UP (817)\n",
      "182\n",
      "    6829 set.\n",
      "    IPA_ID3_DN (817)\n",
      "183\n",
      "    5529 set.\n",
      "    MAHAJAN_RESPONSE_TO_IL1A_UP (78)\n",
      "184\n",
      "    5529 set.\n",
      "    REACTOME_TRISTETRAPROLIN_TTP_ZFP36_BINDS_AND_DESTABILIZES_MRNA (17)\n",
      "185\n",
      "    5529 set.\n",
      "    REACTOME_WNT5A_DEPENDENT_INTERNALIZATION_OF_FZD4 (15)\n",
      "186\n",
      "    5529 set.\n",
      "    REACTOME_POU5F1_OCT4_SOX2_NANOG_ACTIVATE_GENES_RELATED_TO_PROLIFERATION (13)\n",
      "187\n",
      "    5529 set.\n",
      "    BIOCARTA_ERBB3_PATHWAY (5)\n",
      "188\n",
      "    5529 set.\n",
      "    REACTOME_WNT5A_DEPENDENT_INTERNALIZATION_OF_FZD2_FZD5_AND_ROR2 (13)\n",
      "189\n",
      "    15721 set.\n",
      "    GO_CD40_SIGNALING_PATHWAY (12)\n",
      "    BIOCARTA_CD40_PATHWAY (15)\n",
      "19\n",
      "    4762 set.\n",
      "    REACTOME_IL_RECEPTOR_SHC_SIGNALING (27)\n",
      "190\n",
      "    5529 set.\n",
      "    PID_ILK_PATHWAY (45)\n",
      "191\n",
      "    6829 set.\n",
      "    IPA_LEF1_UP (817)\n",
      "192\n",
      "    6829 set.\n",
      "    IPA_IRF7_UP (817)\n",
      "    REACTOME_TRAF6_MEDIATED_IRF7_ACTIVATION_IN_TLR7_8_OR_9_SIGNALING (14)\n",
      "193\n",
      "    5529 set.\n",
      "    REACTOME_FGFR1_LIGAND_BINDING_AND_ACTIVATION (16)\n",
      "194\n",
      "    5529 set.\n",
      "    YAMASHITA_LIVER_CANCER_WITH_EPCAM_UP (53)\n",
      "195\n",
      "    5529 set.\n",
      "    BIOCARTA_GATA3_PATHWAY (14)\n",
      "196\n",
      "    5529 set.\n",
      "    BIOCARTA_CSK_PATHWAY (22)\n",
      "197\n",
      "    5529 set.\n",
      "    ROSS_AML_WITH_CBFB_MYH11_FUSION (52)\n",
      "198\n",
      "    5529 set.\n",
      "    REACTOME_NCAM1_INTERACTIONS (42)\n",
      "199\n",
      "    5529 set.\n",
      "    REACTOME_POLB_DEPENDENT_LONG_PATCH_BASE_EXCISION_REPAIR (8)\n",
      "2\n",
      "    4762 set.\n",
      "    SUMI_HNF4A_TARGETS (34)\n",
      "    LUCAS_HNF4A_TARGETS_UP (58)\n",
      "20\n",
      "    4762 set.\n",
      "    REACTOME_MTORC1_MEDIATED_SIGNALLING (11)\n",
      "    BIOCARTA_IGF1MTOR_PATHWAY (20)\n",
      "    PID_MTOR_4PATHWAY (69)\n",
      "    BIOCARTA_MTOR_PATHWAY (23)\n",
      "    KEGG_MTOR_SIGNALING_PATHWAY (52)\n",
      "200\n",
      "    5529 set.\n",
      "    PID_KIT_PATHWAY (52)\n",
      "201\n",
      "    5529 set.\n",
      "    REACTOME_VLDLR_INTERNALISATION_AND_DEGRADATION (12)\n",
      "202\n",
      "    5529 set.\n",
      "    SETLUR_PROSTATE_CANCER_TMPRSS2_ERG_FUSION_UP (67)\n",
      "203\n",
      "    5529 set.\n",
      "    PLASARI_TGFB1_SIGNALING_VIA_NFIC_1HR_DN (108)\n",
      "204\n",
      "    5529 set.\n",
      "    REACTOME_PTK6_REGULATES_RTKS_AND_THEIR_EFFECTORS_AKT1_AND_DOK1 (9)\n",
      "    REACTOME_PTK6_REGULATES_PROTEINS_INVOLVED_IN_RNA_PROCESSING (5)\n",
      "    REACTOME_PTK6_REGULATES_CELL_CYCLE (6)\n",
      "205\n",
      "    5579 set.\n",
      "    SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP (48)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_UP (83)\n",
      "206\n",
      "    6829 set.\n",
      "    IPA_STAT5A_UP (817)\n",
      "    WIERENGA_STAT5A_TARGETS_DN (193)\n",
      "207\n",
      "    5529 set.\n",
      "    RUIZ_TNC_TARGETS_UP (156)\n",
      "208\n",
      "    5529 set.\n",
      "    REACTOME_RUNX2_REGULATES_GENES_INVOLVED_IN_CELL_MIGRATION (8)\n",
      "    REACTOME_RUNX2_REGULATES_CHONDROCYTE_MATURATION (5)\n",
      "209\n",
      "    5529 set.\n",
      "    REACTOME_SIGNALING_BY_PDGFRA_TRANSMEMBRANE_JUXTAMEMBRANE_AND_KINASE_DOMAIN_MUTANTS (12)\n",
      "21\n",
      "    4762 set.\n",
      "    BIOCARTA_RB_PATHWAY (13)\n",
      "    BIOCARTA_IL2RB_PATHWAY (38)\n",
      "    EGUCHI_CELL_CYCLE_RB1_TARGETS (23)\n",
      "210\n",
      "    5529 set.\n",
      "    REACTOME_FGFR2_ALTERNATIVE_SPLICING (27)\n",
      "211\n",
      "    5529 set.\n",
      "    KAMIKUBO_MYELOID_CEBPA_NETWORK (30)\n",
      "212\n",
      "    5529 set.\n",
      "    WAKASUGI_HAVE_ZNF143_BINDING_SITES (54)\n",
      "213\n",
      "    10192 set.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "    GO_LSM1_7_PAT1_COMPLEX (5)\n",
      "214\n",
      "    6829 set.\n",
      "    IPA_MITF_UP (817)\n",
      "    IPA_MITF_DN (817)\n",
      "215\n",
      "    11701 set.\n",
      "    IPA_STAT6_DN (817)\n",
      "    GSE25088_IL4_VS_IL4_AND_ROSIGLITAZONE_STIM_STAT6_KO_MACROPHAGE_DAY10_UP (197)\n",
      "216\n",
      "    5529 set.\n",
      "    REACTOME_TFAP2A_ACTS_AS_A_TRANSCRIPTIONAL_REPRESSOR_DURING_RETINOIC_ACID_INDUCED_CELL_DIFFERENTIATION (5)\n",
      "217\n",
      "    5718 set.\n",
      "    IWANAGA_CARCINOGENESIS_BY_KRAS_UP (173)\n",
      "    SWEET_KRAS_TARGETS_DN (65)\n",
      "    SINGH_KRAS_DEPENDENCY_SIGNATURE (20)\n",
      "218\n",
      "    6829 set.\n",
      "    IPA_GLI2_UP (817)\n",
      "219\n",
      "    5529 set.\n",
      "    RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_SMAD4_UP (10)\n",
      "22\n",
      "    4762 set.\n",
      "    KIM_PTEN_TARGETS_DN (6)\n",
      "    SA_PTEN_PATHWAY (17)\n",
      "    BIOCARTA_PTEN_PATHWAY (18)\n",
      "220\n",
      "    5529 set.\n",
      "    BILD_HRAS_ONCOGENIC_SIGNATURE (249)\n",
      "221\n",
      "    6829 set.\n",
      "    IPA_BHLHE41_DN (817)\n",
      "222\n",
      "    5529 set.\n",
      "    JOHNSTONE_PARVB_TARGETS_1_DN (62)\n",
      "223\n",
      "    5529 set.\n",
      "    REACTOME_SUMO_IS_CONJUGATED_TO_E1_UBA2_SAE1 (5)\n",
      "224\n",
      "    6829 set.\n",
      "    IPA_AHR_UP (817)\n",
      "225\n",
      "    5529 set.\n",
      "    REACTOME_MISSPLICED_LRP5_MUTANTS_HAVE_ENHANCED_BETA_CATENIN_DEPENDENT_SIGNALING (6)\n",
      "226\n",
      "    6829 set.\n",
      "    IPA_MYBL2_DN (817)\n",
      "227\n",
      "    11701 set.\n",
      "    GSE36891_UNSTIM_VS_POLYIC_TLR3_STIM_PERITONEAL_MACROPHAGE_UP (156)\n",
      "228\n",
      "    11701 set.\n",
      "    GSE34156_UNTREATED_VS_6H_TLR1_TLR2_LIGAND_TREATED_MONOCYTE_DN (179)\n",
      "    GSE34156_UNTREATED_VS_6H_TLR1_TLR2_LIGAND_TREATED_MONOCYTE_UP (196)\n",
      "    GSE34156_TLR1_TLR2_LIGAND_VS_NOD2_AND_TLR1_TLR2_LIGAND_24H_TREATED_MONOCYTE_UP (194)\n",
      "229\n",
      "    5529 set.\n",
      "    REACTOME_NR1H2_NR1H3_REGULATE_GENE_EXPRESSION_TO_LIMIT_CHOLESTEROL_UPTAKE (5)\n",
      "    REACTOME_NR1H2_NR1H3_REGULATE_GENE_EXPRESSION_LINKED_TO_TRIGLYCERIDE_LIPOLYSIS_IN_ADIPOSE (5)\n",
      "23\n",
      "    4812 set.\n",
      "    PID_PI3KCI_PATHWAY (49)\n",
      "    PID_IL2_PI3K_PATHWAY (34)\n",
      "230\n",
      "    10192 set.\n",
      "    GO_MON1_CCZ1_COMPLEX (5)\n",
      "231\n",
      "    5529 set.\n",
      "    PID_FOXM1_PATHWAY (40)\n",
      "232\n",
      "    5529 set.\n",
      "    REACTOME_REGULATION_OF_TP53_ACTIVITY_THROUGH_ASSOCIATION_WITH_CO_FACTORS (14)\n",
      "    REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_CASPASE_ACTIVATORS_AND_CASPASES (12)\n",
      "233\n",
      "    5529 set.\n",
      "    BIOCARTA_PITX2_PATHWAY (16)\n",
      "234\n",
      "    11701 set.\n",
      "    IPA_STAT1_UP (817)\n",
      "    GSE40666_UNTREATED_VS_IFNA_STIM_STAT1_KO_CD8_TCELL_90MIN_UP (188)\n",
      "235\n",
      "    5529 set.\n",
      "    REACTOME_L1CAM_INTERACTIONS (121)\n",
      "236\n",
      "    5529 set.\n",
      "    REACTOME_CHL1_INTERACTIONS (9)\n",
      "237\n",
      "    5529 set.\n",
      "    HOOI_ST7_TARGETS_DN (118)\n",
      "    HOOI_ST7_TARGETS_UP (92)\n",
      "238\n",
      "    5529 set.\n",
      "    REACTOME_NEGATIVE_REGULATORS_OF_DDX58_IFIH1_SIGNALING (34)\n",
      "239\n",
      "    5529 set.\n",
      "    HUANG_GATA2_TARGETS_DN (74)\n",
      "24\n",
      "    10679 set.\n",
      "    KEGG_FOCAL_ADHESION (201)\n",
      "    GO_CELL_MATRIX_ADHESION (119)\n",
      "    GO_CELL_SUBSTRATE_ADHESION (164)\n",
      "240\n",
      "    6829 set.\n",
      "    IPA_TCF12_UP (817)\n",
      "241\n",
      "    5529 set.\n",
      "    REACTOME_DEFECTIVE_CHSY1_CAUSES_TPBS (8)\n",
      "242\n",
      "    5529 set.\n",
      "    REACTOME_SEMA3A_PLEXIN_REPULSION_SIGNALING_BY_INHIBITING_INTEGRIN_ADHESION (14)\n",
      "243\n",
      "    5529 set.\n",
      "    REACTOME_FGFRL1_MODULATION_OF_FGFR1_SIGNALING (13)\n",
      "244\n",
      "    10401 set.\n",
      "    GSE25085_FETAL_LIVER_VS_ADULT_BM_SP4_THYMIC_IMPLANT_DN (200)\n",
      "245\n",
      "    4951 set.\n",
      "    WNT_UP.V1_UP (180)\n",
      "    KEGG_MAPK_SIGNALING_PATHWAY (267)\n",
      "    VERHAAK_AML_WITH_NPM1_MUTATED_DN (246)\n",
      "    BIOCARTA_CELLCYCLE_PATHWAY (23)\n",
      "    MYC_UP.V1_UP (186)\n",
      "246\n",
      "    4762 set.\n",
      "    VERHAAK_GLIOBLASTOMA_PRONEURAL (177)\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL (216)\n",
      "    VERHAAK_GLIOBLASTOMA_CLASSICAL (162)\n",
      "    PID_WNT_CANONICAL_PATHWAY (20)\n",
      "    PID_NOTCH_PATHWAY (59)\n",
      "247\n",
      "    4762 set.\n",
      "    YU_MYC_TARGETS_UP (42)\n",
      "248\n",
      "    5529 set.\n",
      "    REACTOME_GLUCOSE_METABOLISM (92)\n",
      "    REACTOME_GLYCOLYSIS (72)\n",
      "    KEGG_PYRUVATE_METABOLISM (40)\n",
      "    KEGG_GLYCOLYSIS_GLUCONEOGENESIS (62)\n",
      "249\n",
      "    5529 set.\n",
      "    HSIAO_HOUSEKEEPING_GENES (399)\n",
      "    WONG_EMBRYONIC_STEM_CELL_CORE (335)\n",
      "    REACTOME_TRANSLATION (295)\n",
      "    RHEIN_ALL_GLUCOCORTICOID_THERAPY_DN (368)\n",
      "    REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE (113)\n",
      "    SPIELMAN_LYMPHOBLAST_EUROPEAN_VS_ASIAN_UP (484)\n",
      "25\n",
      "    4762 set.\n",
      "    REACTOME_ACTIVATED_TAK1_MEDIATES_P38_MAPK_ACTIVATION (18)\n",
      "    BIOCARTA_P38MAPK_PATHWAY (40)\n",
      "250\n",
      "    5529 set.\n",
      "    BROWNE_INTERFERON_RESPONSIVE_GENES (67)\n",
      "    REACTOME_CHOLESTEROL_BIOSYNTHESIS (25)\n",
      "251\n",
      "    6278 set.\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL (216)\n",
      "    VERHAAK_GLIOBLASTOMA_PRONEURAL (177)\n",
      "    PHILLIPS_GLIOBLASTOMA_MESENCHYMAL (572)\n",
      "    PHILLIPS__GLIOBLASTOMA_PRONEURAL (572)\n",
      "    ANASTASSIOU_MULTICANCER_INVASIVENESS (572)\n",
      "    SEGERMAN_MULTITHERAPY_RESISTANCE (572)\n",
      "252\n",
      "    14765 set.\n",
      "    GO_RIBOSOME (253)\n",
      "    GO_CYTOKINE_ACTIVITY (234)\n",
      "    GO_MATURE_RIBOSOME_ASSEMBLY (6)\n",
      "    GO_CYTOSOLIC_RIBOSOME (110)\n",
      "    GO_ENDOTHELIAL_CELL_MATRIX_ADHESION (10)\n",
      "    GO_REGULATION_OF_CHAPERONE_MEDIATED_PROTEIN_FOLDING (5)\n",
      "    GO_PROTEIN_FOLDING_IN_ENDOPLASMIC_RETICULUM (10)\n",
      "253\n",
      "    6279 set.\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP (202)\n",
      "254\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    HALLMARK_P53_PATHWAY (200)\n",
      "255\n",
      "    6279 set.\n",
      "    KEGG_ECM_RECEPTOR_INTERACTION (84)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "256\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    PID_E2F_PATHWAY (73)\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "    REACTOME_CELL_CYCLE (693)\n",
      "    REACTOME_CELL_CYCLE_MITOTIC (561)\n",
      "257\n",
      "    6279 set.\n",
      "    HALLMARK_CHOLESTEROL_HOMEOSTASIS (74)\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "258\n",
      "    6229 set.\n",
      "    MOSERLE_IFNA_RESPONSE (31)\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (71)\n",
      "259\n",
      "    11101 set.\n",
      "    GSE14308_TH17_VS_NATURAL_TREG_DN (200)\n",
      "26\n",
      "    10679 set.\n",
      "    BOSCO_EPITHELIAL_DIFFERENTIATION_MODULE (69)\n",
      "    GO_EPITHELIAL_CELL_DIFFERENTIATION (495)\n",
      "260\n",
      "    6279 set.\n",
      "    HALLMARK_SPERMATOGENESIS (135)\n",
      "    HALLMARK_MTORC1_SIGNALING (200)\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "261\n",
      "    6279 set.\n",
      "    HALLMARK_XENOBIOTIC_METABOLISM (200)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "    HALLMARK_COAGULATION (138)\n",
      "262\n",
      "    21233 set.\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION (200)\n",
      "    GO_POSITIVE_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION (50)\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE (57)\n",
      "    SARRIO_EPITHELIAL_MESENCHYMAL_TRANSITION_UP (166)\n",
      "263\n",
      "    6229 set.\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL (216)\n",
      "    LIM_MAMMARY_LUMINAL_PROGENITOR_UP (58)\n",
      "    JECHLINGER_EPITHELIAL_TO_MESENCHYMAL_TRANSITION_UP (68)\n",
      "264\n",
      "    11151 set.\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY (49)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "    HALLMARK_APOPTOSIS (161)\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_PRIMARY_INF_CD8_TCELL_UP (200)\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_PRIMARY_INF_CD8_TCELL_DN (200)\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_SECONDARY_INF_CD8_TCELL_UP (200)\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_SECONDARY_INF_CD8_TCELL_DN (200)\n",
      "    GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_UP (200)\n",
      "    GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_DN (200)\n",
      "265\n",
      "    6328 set.\n",
      "    BIDUS_METASTASIS_UP (219)\n",
      "    CRC_proliferation_signature (572)\n",
      "    BIDUS_METASTASIS_DN (154)\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION (200)\n",
      "266\n",
      "    11151 set.\n",
      "    GSE9650_EFFECTOR_VS_MEMORY_CD8_TCELL_UP (199)\n",
      "267\n",
      "    6279 set.\n",
      "    HALLMARK_MTORC1_SIGNALING (200)\n",
      "268\n",
      "    6279 set.\n",
      "    KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION (88)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "27\n",
      "    4762 set.\n",
      "    REACTOME_CYCLIN_A_B1_ASSOCIATED_EVENTS_DURING_G2_M_TRANSITION (15)\n",
      "    CHANG_CYCLING_GENES (148)\n",
      "    REACTOME_APC_C_CDC20_MEDIATED_DEGRADATION_OF_CYCLIN_B (26)\n",
      "28\n",
      "    4762 set.\n",
      "    PID_TCR_JNK_PATHWAY (14)\n",
      "    HAN_JNK_SINGALING_UP (35)\n",
      "    REACTOME_JNK_C_JUN_KINASES_PHOSPHORYLATION_AND_ACTIVATION_MEDIATED_BY_ACTIVATED_HUMAN_TAK1 (16)\n",
      "29\n",
      "    4762 set.\n",
      "    BIOCARTA_GATA3_PATHWAY (16)\n",
      "3\n",
      "    4951 set.\n",
      "    NFE2L2.V2 (481)\n",
      "    SINGH_NFE2L2_TARGETS (15)\n",
      "30\n",
      "    4762 set.\n",
      "    BAKKER_FOXO3_TARGETS_UP (61)\n",
      "31\n",
      "    4762 set.\n",
      "    PID_FOXM1_PATHWAY (40)\n",
      "32\n",
      "    4762 set.\n",
      "    REACTOME_DEGRADATION_OF_THE_EXTRACELLULAR_MATRIX (29)\n",
      "    REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION (87)\n",
      "    REACTOME_CELL_EXTRACELLULAR_MATRIX_INTERACTIONS (14)\n",
      "    BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX (17)\n",
      "    REACTOME_P130CAS_LINKAGE_TO_MAPK_SIGNALING_FOR_INTEGRINS (15)\n",
      "33\n",
      "    4762 set.\n",
      "    VANTVEER_BREAST_CANCER_ESR1 (missing)\n",
      "    DOANE_BREAST_CANCER_ESR1 (missing)\n",
      "    YANG_BREAST_CANCER_ESR1_UP (36)\n",
      "34\n",
      "    4762 set.\n",
      "    BIOCARTA_CASPASE_PATHWAY (23)\n",
      "    SA_CASPASE_CASCADE (19)\n",
      "    PID_CASPASE_PATHWAY (52)\n",
      "    REACTOME_NFKB_ACTIVATION_THROUGH_FADD_RIP1_PATHWAY_MEDIATED_BY_CASPASE_8_AND10 (12)\n",
      "    REACTOME_CASPASE_MEDIATED_CLEAVAGE_OF_CYTOSKELETAL_PROTEINS (13)\n",
      "35\n",
      "    5001 set.\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP (47)\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    DANG_REGULATED_BY_MYC (missing)\n",
      "    MYC_UP.V1_UP (186)\n",
      "    ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS (missing)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS (missing)\n",
      "    PID_MYC_ACTIV_PATHWAY (79)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    DANG_BOUND_BY_MYC (1103)\n",
      "    SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM (159)\n",
      "    BENPORATH_MYC_MAX_TARGETS (775)\n",
      "    MYC_UP.V1 (missing)\n",
      "    COLLER_MYC_TARGETS (missing)\n",
      "    ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_UP (84)\n",
      "36\n",
      "    4762 set.\n",
      "    SEIDEN_ONCOGENESIS_BY_MET (88)\n",
      "37\n",
      "    4762 set.\n",
      "    REACTOME_REGULATION_OF_RHEB_GTPASE_ACTIVITY_BY_AMPK (10)\n",
      "    REACTOME_REGULATION_OF_AMPK_ACTIVITY_VIA_LKB1 (15)\n",
      "    REACTOME_ENERGY_DEPENDENT_REGULATION_OF_MTOR_BY_LKB1_AMPK (18)\n",
      "    REACTOME_ACTIVATED_AMPK_STIMULATES_FATTY_ACID_OXIDATION_IN_MUSCLE (19)\n",
      "38\n",
      "    10679 set.\n",
      "    GO_TRANSLATION_INITIATION_FACTOR_ACTIVITY (51)\n",
      "    GO_REGULATORY_REGION_NUCLEIC_ACID_BINDING (818)\n",
      "    GO_TRANSLATION_REGULATOR_ACTIVITY_NUCLEIC_ACID_BINDING (18)\n",
      "    GO_EUKARYOTIC_TRANSLATION_INITIATION_FACTOR_3_COMPLEX (16)\n",
      "    GO_TRANSLATIONAL_INITIATION (146)\n",
      "    REACTOME_TRANSLATION (222)\n",
      "    GO_TRANSLATION_REGULATOR_ACTIVITY (35)\n",
      "    GO_REGULATION_OF_TRANSLATIONAL_INITIATION (82)\n",
      "39\n",
      "    4812 set.\n",
      "    NELSON_RESPONSE_TO_ANDROGEN_UP (86)\n",
      "    HALLMARK_ANDROGEN_RESPONSE (101)\n",
      "    DOANE_RESPONSE_TO_ANDROGEN_UP (184)\n",
      "4\n",
      "    4762 set.\n",
      "    REACTOME_SIGNALING_BY_ACTIVATED_POINT_MUTANTS_OF_FGFR1 (11)\n",
      "    REACTOME_FGFR1_LIGAND_BINDING_AND_ACTIVATION (14)\n",
      "40\n",
      "    4762 set.\n",
      "    PUJANA_ATM_PCC_NETwork (missing)\n",
      "    PID_ATM_PATHWAY (34)\n",
      "41\n",
      "    189 set.\n",
      "    EGFR_UP.V1_UP (193)\n",
      "42\n",
      "    10729 set.\n",
      "    HALLMARK_ANGIOGENESIS (36)\n",
      "    LU_TUMOR_ANGIOGENESIS_UP (25)\n",
      "    GO_ANGIOGENESIS (293)\n",
      "43\n",
      "    10679 set.\n",
      "    BIOCARTA_TGFB_PATHWAY (19)\n",
      "    GO_REGULATION_OF_STEM_CELL_PROLIFERATION (88)\n",
      "    REACTOME_SIGNALING_BY_NODAL (18)\n",
      "44\n",
      "    6229 set.\n",
      "    REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION (301)\n",
      "    PID_A6B1_A6B4_INTEGRIN_PATHWAY (46)\n",
      "    NABA_CORE_MATRISOME (275)\n",
      "    KEGG_ECM_RECEPTOR_INTERACTION (84)\n",
      "45\n",
      "    5917 set.\n",
      "    GO_PHAGOSOME_MATURATION (38)\n",
      "    GO_PHAGOSOME_ACIDIFICATION (27)\n",
      "46\n",
      "    4762 set.\n",
      "    BOQUEST_STEM_CELL_UP (260)\n",
      "    SCHUETZ_BREAST_CANCER_DUCTAL_INVASIVE_UP (351)\n",
      "    LIM_MAMMARY_STEM_CELL_UP (489)\n",
      "47\n",
      "    189 set.\n",
      "    ALK_DN.V1_UP (145)\n",
      "48\n",
      "    5917 set.\n",
      "    GO_SKELETAL_SYSTEM_DEVELOPMENT (455)\n",
      "49\n",
      "    6279 set.\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    WONG_EMBRYONIC_STEM_CELL_CORE (335)\n",
      "5\n",
      "    4762 set.\n",
      "    PID_ALK2_PATHWAY (11)\n",
      "    TESAR_ALK_TARGETS_EPISC_4D_UP (7)\n",
      "    TESAR_ALK_TARGETS_EPISC_3D_UP (7)\n",
      "50\n",
      "    50 set.\n",
      "    HALLMARK_CHOLESTEROL_HOMEOSTASIS (74)\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE (200)\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE (97)\n",
      "51\n",
      "    4762 set.\n",
      "    FEVR_CTNNB1_TARGETS_DN (553)\n",
      "    HUPER_BREAST_BASAL_VS_LUMINAL_UP (54)\n",
      "    BENPORATH_ES_1 (379)\n",
      "    SHEDDEN_LUNG_CANCER_POOR_SURVIVAL_A6 (456)\n",
      "    NADERI_BREAST_CANCER_PROGNOSIS_UP (50)\n",
      "52\n",
      "    4811 set.\n",
      "    NRAS_EXTINCTION_DOWN (572)\n",
      "    NEBEN_AML_WITH_FLT3_OR_NRAS_DN (12)\n",
      "53\n",
      "    4762 set.\n",
      "    REACTOME_CELL_CYCLE (421)\n",
      "    PID_E2F_PATHWAY (74)\n",
      "    REACTOME_DNA_REPLICATION (192)\n",
      "54\n",
      "    10728 set.\n",
      "    GO_SEROTONIN_RECEPTOR_SIGNALING_PATHWAY (572)\n",
      "    DIERICK_SEROTONIN_FUNCTION_GENES (6)\n",
      "    REACTOME_SEROTONIN_RECEPTORS (12)\n",
      "    GO_G_PROTEIN_COUPLED_RECEPTOR_SIGNALING_PATHWAY (1193)\n",
      "55\n",
      "    5050 set.\n",
      "    ERBB2_UP.V1_UP (572)\n",
      "    SMID_BREAST_CANCER_ERBB2_UP (147)\n",
      "    MEK_UP.V1_UP (196)\n",
      "    LIM_MAMMARY_LUMINAL_MATURE_UP (116)\n",
      "    HINATA_NFKB_TARGETS_KERATINOCYTE_UP (91)\n",
      "    HINATA_NFKB_TARGETS_FIBROBLAST_UP (84)\n",
      "56\n",
      "    4762 set.\n",
      "    BAKER_HEMATOPOESIS_STAT1_TARGETS (10)\n",
      "    DER_IFN_GAMMA_RESPONSE (missing)\n",
      "57\n",
      "    4812 set.\n",
      "    INGA_TP53_TARGETS (17)\n",
      "    KIM_HYPOXIA (25)\n",
      "    LEONARD_HYPOXIA (47)\n",
      "58\n",
      "    4861 set.\n",
      "    REACTOME_PI3K_AKT_ACTIVATION (38)\n",
      "    BIOCARTA_GSK3_PATHWAY (27)\n",
      "59\n",
      "    50 set.\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "6\n",
      "    4811 set.\n",
      "    GROGER_EMT (missing)\n",
      "    TAUBE_EMT (missing)\n",
      "60\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_OXIDATIVE_PHOSPHORYLATION (200)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    HALLMARK_DNA_REPAIR (150)\n",
      "61\n",
      "    6229 set.\n",
      "    REACTOME_TOLL_LIKE_RECEPTOR_CASCADES (153)\n",
      "    NEMETH_INFLAMMATORY_RESPONSE_LPS_UP (85)\n",
      "    KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY (102)\n",
      "    SEKI_INFLAMMATORY_RESPONSE_LPS_UP (78)\n",
      "62\n",
      "    10729 set.\n",
      "    HALLMARK_ANDROGEN_RESPONSE (101)\n",
      "    NELSON_RESPONSE_TO_ANDROGEN_UP (86)\n",
      "    DOANE_RESPONSE_TO_ANDROGEN_UP (184)\n",
      "    WANG_RESPONSE_TO_ANDROGEN_UP (29)\n",
      "63\n",
      "    4762 set.\n",
      "    KEGG_AMINOACYL_TRNA_BIOSYNTHESIS (41)\n",
      "    KEGG_PYRIMIDINE_METABOLISM (98)\n",
      "64\n",
      "    4811 set.\n",
      "    MAMMOSPHERES (572)\n",
      "65\n",
      "    5969 set.\n",
      "    ROS_GENES (572)\n",
      "66\n",
      "    14765 set.\n",
      "    GO_T_CELL_DIFFERENTIATION (253)\n",
      "    GO_REGULATION_OF_LEUKOCYTE_CHEMOTAXIS (124)\n",
      "    GO_LEUKOCYTE_CELL_CELL_ADHESION (364)\n",
      "    GO_ADAPTIVE_IMMUNE_RESPONSE (687)\n",
      "    GO_POSITIVE_REGULATION_OF_T_CELL_PROLIFERATION (101)\n",
      "    GO_DEFENSE_RESPONSE_TO_BACTERIUM (341)\n",
      "    GO_INFLAMMATORY_RESPONSE (801)\n",
      "67\n",
      "    4762 set.\n",
      "    REACTOME_DNA_REPLICATION (192)\n",
      "    KEGG_DNA_REPLICATION (36)\n",
      "68\n",
      "    4762 set.\n",
      "    REACTOME_TRANSCRIPTION (210)\n",
      "    REACTOME_TRANSLATION (222)\n",
      "    KEGG_DNA_REPLICATION (36)\n",
      "69\n",
      "    6229 set.\n",
      "    PUJANA_BREAST_CANCER_WITH_BRCA1_MUTATED_UP (55)\n",
      "    COLLER_MYC_TARGETS_UP (24)\n",
      "7\n",
      "    4762 set.\n",
      "    KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS (28)\n",
      "    KEGG_PENTOSE_PHOSPHATE_PATHWAY (27)\n",
      "70\n",
      "    4762 set.\n",
      "    KEGG_GLYCOLYSIS_GLUCONEOGENESIS (62)\n",
      "    BIOCARTA_GLYCOLYSIS_PATHWAY (10)\n",
      "    REACTOME_GLYCOLYSIS (29)\n",
      "    MOOTHA_GLYCOLYSIS (21)\n",
      "71\n",
      "    6229 set.\n",
      "    LEE_BMP2_TARGETS_UP (773)\n",
      "    NABA_MATRISOME (1026)\n",
      "    NABA_MATRISOME_ASSOCIATED (751)\n",
      "    NABA_CORE_MATRISOME (275)\n",
      "    KEGG_LYSOSOME (121)\n",
      "72\n",
      "    50 set.\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "73\n",
      "    4762 set.\n",
      "    PENG_RAPAMYCIN_RESPONSE_DN (245)\n",
      "74\n",
      "    5917 set.\n",
      "    GO_ANTIGEN_BINDING (114)\n",
      "    GO_ALPHA_BETA_T_CELL_ACTIVATION (54)\n",
      "75\n",
      "    5648 set.\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    MENSSEN_MYC_TARGETS (53)\n",
      "    SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP (47)\n",
      "76\n",
      "    4762 set.\n",
      "    ZHANG_TARGETS_OF_EWSR1_FLI1_FUSION (88)\n",
      "    RIGGI_EWING_SARCOMA_PROGENITOR_UP (430)\n",
      "    STAEGE_EWING_FAMILY_TUMOR (33)\n",
      "77\n",
      "    4762 set.\n",
      "    KONG_E2F3_TARGETS (97)\n",
      "    REACTOME_CYCLIN_A_B1_ASSOCIATED_EVENTS_DURING_G2_M_TRANSITION (15)\n",
      "    WHITFIELD_CELL_CYCLE_G2 (182)\n",
      "    REACTOME_GENERIC_TRANSCRIPTION_PATHWAY (352)\n",
      "78\n",
      "    4762 set.\n",
      "    KEGG_STEROID_BIOSYNTHESIS (17)\n",
      "    KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS (16)\n",
      "    KEGG_PEROXISOME (78)\n",
      "79\n",
      "    4762 set.\n",
      "    KEGG_LYSOSOME (121)\n",
      "    KEGG_REGULATION_OF_AUTOPHAGY (35)\n",
      "    KEGG_FATTY_ACID_METABOLISM (42)\n",
      "    KEGG_PEROXISOME (78)\n",
      "    REACTOME_RESPIRATORY_ELECTRON_TRANSPORT (79)\n",
      "8\n",
      "    4762 set.\n",
      "    VANTVEER_BREAST_CANCER_ESR1 (missing)\n",
      "    DOANE_BREAST_CANCER_ESR1 (missing)\n",
      "80\n",
      "    239 set.\n",
      "    HALLMARK_KRAS_SIGNALING_UP (200)\n",
      "    EGFR_UP.V1 (missing)\n",
      "    KRAS.LUNG.BREAST_UP.V1 (missing)\n",
      "81\n",
      "    6229 set.\n",
      "    MILI_PSEUDOPODIA_HAPTOTAXIS_UP (522)\n",
      "    HAMAI_APOPTOSIS_VIA_TRAIL_UP (655)\n",
      "    SEIDEN_ONCOGENESIS_BY_MET (89)\n",
      "82\n",
      "    4812 set.\n",
      "    BILD_MYC_ONCOGENIC_SIGNATURE (206)\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    HALLMARK_MYC_TARGETS_V1 (200)\n",
      "    HALLMARK_MYC_TARGETS_V2 (58)\n",
      "    COLLER_MYC_TARGETS_UP (25)\n",
      "83\n",
      "    4762 set.\n",
      "    DANG_REGULATED_BY_MYC_UP (72)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP (201)\n",
      "    SCHUHMACHER_MYC_TARGETS (missing)\n",
      "    COLLER_MYC_TARGETS (missing)\n",
      "    YU_MYC_TARGETS (missing)\n",
      "    DANG_REGULATED_BY_MYC (missing)\n",
      "84\n",
      "    4762 set.\n",
      "    ZHU_CMV_8_HR_UP (47)\n",
      "    DER_IFN_ALPHA_RESPONSE_UP (74)\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (64)\n",
      "    EINAV_INTERFERON_SIGNATURE_IN_CANCER (27)\n",
      "    BROWNE_INTERFERON_RESPONSIVE_GENES (68)\n",
      "85\n",
      "    4762 set.\n",
      "    WNT_SIGNALING (89)\n",
      "    YU_MYC_TARGETS_UP (42)\n",
      "    KEGG_CELL_CYCLE (128)\n",
      "    REACTOME_CELL_CYCLE (421)\n",
      "86\n",
      "    4762 set.\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (64)\n",
      "    ZHANG_INTERFERON_RESPONSE (23)\n",
      "    EINAV_INTERFERON_SIGNATURE_IN_CANCER (27)\n",
      "    ST_INTERFERON_GAMMA_PATHWAY (10)\n",
      "87\n",
      "    6328 set.\n",
      "    activated_treg (572)\n",
      "    HALLMARK_IL2_STAT5_SIGNALING (199)\n",
      "    HALLMARK_G2M_CHECKPOINT (200)\n",
      "    HALLMARK_E2F_TARGETS (200)\n",
      "88\n",
      "    4762 set.\n",
      "    SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY (32)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP (201)\n",
      "    COLLER_MYC_TARGETS_UP (25)\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    BILANGES_SERUM_AND_RAPAMYCIN_SENSITIVE_GENES (68)\n",
      "89\n",
      "    20994 set.\n",
      "    KEGG_HOMOLOGOUS_RECOMBINATION (28)\n",
      "    GO_MITOTIC_NUCLEAR_DIVISION (277)\n",
      "    GO_REGULATION_OF_TRANSCRIPTION_INVOLVED_IN_G1_S_TRANSITION_OF_MITOTIC_CELL_CYCLE (35)\n",
      "9\n",
      "    4812 set.\n",
      "    HINATA_NFKB_TARGETS_FIBROBLAST_UP (84)\n",
      "    SCHOEN_NFKB_SIGNALING (34)\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB (200)\n",
      "    TIAN_TNF_SIGNALING_VIA_NFKB (28)\n",
      "90\n",
      "    4762 set.\n",
      "    SCHUHMACHER_MYC_TARGETS_UP (80)\n",
      "    DANG_MYC_TARGETS_UP (143)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS (missing)\n",
      "    CAIRO_PML_TARGETS_BOUND_BY_MYC_UP (23)\n",
      "91\n",
      "    4811 set.\n",
      "    LSC_ASSOCIATED_GENES (572)\n",
      "92\n",
      "    5917 set.\n",
      "    GO_REGULATION_OF_MONONUCLEAR_CELL_MIGRATION (16)\n",
      "93\n",
      "    4811 set.\n",
      "    Amino_Acid_Metabolism (572)\n",
      "94\n",
      "    4811 set.\n",
      "    Erythroid_associated_genes (572)\n",
      "95\n",
      "    6111 set.\n",
      "    AML_TCP_vs_untreated_UP (572)\n",
      "96\n",
      "    14765 set.\n",
      "    GO_DNA_REPLICATION (273)\n",
      "    GO_DNA_DEPENDENT_DNA_REPLICATION (151)\n",
      "    GO_REGULATION_OF_DNA_DEPENDENT_DNA_REPLICATION (52)\n",
      "    GO_DNA_STRAND_ELONGATION (26)\n",
      "    GO_DNA_STRAND_ELONGATION_INVOLVED_IN_DNA_REPLICATION (19)\n",
      "97\n",
      "    5001 set.\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE (57)\n",
      "    HALLMARK_APICAL_JUNCTION (200)\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION (200)\n",
      "98\n",
      "    4811 set.\n",
      "    FOCAL_ADHESION (572)\n",
      "    LIANG_SILENCED_BY_METHYLATION (missing)\n",
      "    MISSIAGLIA_REGULATED_BY_METHYLATION_UP (126)\n",
      "99\n",
      "    4762 set.\n",
      "    KEGG_SPLICEOSOME (128)\n",
      "    KEGG_RNA_DEGRADATION (59)\n",
      "    REACTOME_P53_DEPENDENT_G1_DNA_DAMAGE_RESPONSE (57)\n",
      "    REACTOME_P53_INDEPENDENT_G1_S_DNA_DAMAGE_CHECKPOINT (51)\n",
      "    REACTOME_REPAIR_SYNTHESIS_FOR_GAP_FILLING_BY_DNA_POL_IN_TC_NER (14)\n"
     ]
    }
   ],
   "source": [
    "for benchmark_directory_path in read_directory(benchmarks_directory_path)\n",
    "\n",
    "    #\n",
    "    name = splitpath(benchmark_directory_path)[end]\n",
    "\n",
    "    println(name)\n",
    "\n",
    "    #\n",
    "    element_, score_ =\n",
    "        eachcol(read_data(joinpath(benchmark_directory_path, \"gene_x_score.tsv\")))\n",
    "\n",
    "    #\n",
    "    json_dict = JSON.parse(open(joinpath(benchmark_directory_path, \"gene_sets.json\")))\n",
    "\n",
    "    #\n",
    "    set_to_element_ = Dict{String,Vector{String}}()\n",
    "\n",
    "    for gmt_path in json_dict[\"gene_sets_tested\"]\n",
    "\n",
    "        merge!(set_to_element_, name_to_set_to_element_[splitpath(gmt_path)[end]])\n",
    "\n",
    "    end\n",
    "\n",
    "    set_ = sort(collect(keys(set_to_element_)))\n",
    "\n",
    "    n_set = length(set_)\n",
    "\n",
    "    println(\"    \", n_set, \" set.\")\n",
    "\n",
    "    #\n",
    "    set_to_method_to_result = sort(score_set_new(element_, score_, set_to_element_))\n",
    "\n",
    "    @assert all(set_ .== collect(keys(set_to_method_to_result)))\n",
    "\n",
    "    set_x_method = Matrix{Float64}(undef, n_set, n_method)\n",
    "\n",
    "    for (set_i, (set, method_to_result)) in enumerate(set_to_method_to_result)\n",
    "\n",
    "        @assert set_[set_i] == set\n",
    "\n",
    "        set_x_method_row = []\n",
    "\n",
    "        for result in values(method_to_result)\n",
    "\n",
    "            append!(set_x_method_row, result[2:3])\n",
    "\n",
    "        end\n",
    "\n",
    "        set_x_method[set_i, :] = set_x_method_row\n",
    "\n",
    "    end\n",
    "\n",
    "    #\n",
    "    for set in json_dict[\"gene_sets_positive\"]\n",
    "\n",
    "        #\n",
    "        print(\"    \")\n",
    "\n",
    "        if !in(set, set_)\n",
    "\n",
    "            println(set, \" (missing)\")\n",
    "\n",
    "            continue\n",
    "\n",
    "        end\n",
    "\n",
    "        benchmark = string(name, '.', set)\n",
    "\n",
    "        benchmark_size = length(set_to_element_[set])\n",
    "\n",
    "        println(set, \" (\", benchmark_size, \")\")\n",
    "\n",
    "        #\n",
    "        push!(benchmark_, benchmark)\n",
    "\n",
    "        push!(benchmark_size_, benchmark_size)\n",
    "\n",
    "        #\n",
    "        benchmark_x_method_row = Vector{Float64}()\n",
    "\n",
    "        for (method_i, set_score_) in enumerate(eachcol(set_x_method))\n",
    "\n",
    "            #\n",
    "            is_negative_ = set_score_ .< 0\n",
    "\n",
    "            is_positive_ = 0 .<= set_score_\n",
    "\n",
    "            #\n",
    "            set_negative_ = set_[is_negative_]\n",
    "\n",
    "            set_positive_ = set_[is_positive_]\n",
    "\n",
    "            #\n",
    "            negative_ = .-normalize(.-set_score_[is_negative_], \"1224\") / sum(is_negative_)\n",
    "\n",
    "            positive_ = normalize(set_score_[is_positive_], \"1224\") / sum(is_positive_)\n",
    "\n",
    "            #\n",
    "            i_negative = findfirst(set_negative_ .== set)\n",
    "\n",
    "            i_positive = findfirst(set_positive_ .== set)\n",
    "\n",
    "            #\n",
    "            if i_negative != nothing\n",
    "\n",
    "                benchmark_score = negative_[i_negative]\n",
    "\n",
    "            elseif i_positive != nothing\n",
    "\n",
    "                benchmark_score = positive_[i_positive]\n",
    "\n",
    "            else\n",
    "\n",
    "                benchmark_score = set_score_[findfirst(set_ .== set)]\n",
    "\n",
    "                println(\"        \", method_[method_i], \" returned \", benchmark_score)\n",
    "\n",
    "            end\n",
    "\n",
    "            push!(benchmark_x_method_row, benchmark_score)\n",
    "\n",
    "        end\n",
    "\n",
    "        #\n",
    "        push!(benchmark_x_method_rows, benchmark_x_method_row)\n",
    "\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fa4e580a-1918-45a9-9dc1-a1d66b6212e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>Benchmark</th><th>Is KS &lt; extreme</th></tr><tr><th></th><th>String</th><th>Float64</th></tr></thead><tbody><p>698 rows × 73 columns (omitted printing of 71 columns)</p><tr><th>1</th><td>1.AIGNER_ZEB1_TARGETS</td><td>0.995307</td></tr><tr><th>2</th><td>1.SANSOM_APC_TARGETS_UP</td><td>0.856229</td></tr><tr><th>3</th><td>1.BCAT_GDS748_UP</td><td>0.912969</td></tr><tr><th>4</th><td>1.BCAT.100_UP.V1_UP</td><td>0.956058</td></tr><tr><th>5</th><td>1.PID_WNT_SIGNALING_PATHWAY</td><td>0.393771</td></tr><tr><th>6</th><td>1.LIU_CDX2_TARGETS_UP</td><td>0.938993</td></tr><tr><th>7</th><td>1.LUCAS_HNF4A_TARGETS_UP</td><td>0.957338</td></tr><tr><th>8</th><td>10.LANG_MYB_FAMILY_TARGETS</td><td>0.501956</td></tr><tr><th>9</th><td>10.PID_CMYB_PATHWAY</td><td>0.717719</td></tr><tr><th>10</th><td>100.GO_CELL_CYCLE_G1_S_PHASE_TRANSITION</td><td>0.797134</td></tr><tr><th>11</th><td>100.REACTOME_M_G1_TRANSITION</td><td>0.839927</td></tr><tr><th>12</th><td>100.REACTOME_MITOTIC_M_M_G1_PHASES</td><td>0.889786</td></tr><tr><th>13</th><td>100.KEGG_DNA_REPLICATION</td><td>0.931974</td></tr><tr><th>14</th><td>100.REACTOME_DNA_REPLICATION</td><td>0.880702</td></tr><tr><th>15</th><td>100.EGUCHI_CELL_CYCLE_RB1_TARGETS</td><td>0.99677</td></tr><tr><th>16</th><td>100.WHITFIELD_CELL_CYCLE_LITERATURE</td><td>0.986677</td></tr><tr><th>17</th><td>101.BIOCARTA_NFKB_PATHWAY</td><td>0.866393</td></tr><tr><th>18</th><td>101.TIAN_TNF_SIGNALING_VIA_NFKB</td><td>0.982528</td></tr><tr><th>19</th><td>101.BIOCARTA_TNFR2_PATHWAY</td><td>0.994347</td></tr><tr><th>20</th><td>101.SANA_TNF_SIGNALING_UP</td><td>0.873587</td></tr><tr><th>21</th><td>101.PHONG_TNF_TARGETS_UP</td><td>0.913669</td></tr><tr><th>22</th><td>101.SCHOEN_NFKB_SIGNALING</td><td>0.937307</td></tr><tr><th>23</th><td>101.GILMORE_CORE_NFKB_PATHWAY</td><td>0.957348</td></tr><tr><th>24</th><td>102.REACTOME_ALPHA_LINOLENIC_ACID_ALA_METABOLISM</td><td>0.991628</td></tr><tr><th>25</th><td>102.KEGG_BETA_ALANINE_METABOLISM</td><td>0.912097</td></tr><tr><th>26</th><td>103.REACTOME_CELL_CYCLE</td><td>0.877481</td></tr><tr><th>27</th><td>103.EGUCHI_CELL_CYCLE_RB1_TARGETS</td><td>0.990458</td></tr><tr><th>28</th><td>103.REACTOME_REGULATION_OF_MITOTIC_CELL_CYCLE</td><td>0.925191</td></tr><tr><th>29</th><td>104.ESW_FLI_UPREGULATED</td><td>0.85987</td></tr><tr><th>30</th><td>105.REACTOME_THE_CITRIC_ACID_TCA_CYCLE_AND_RESPIRATORY_ELECTRON_TRANSPORT</td><td>0.733066</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ccc}\n",
       "\t& Benchmark & Is KS < extreme & \\\\\n",
       "\t\\hline\n",
       "\t& String & Float64 & \\\\\n",
       "\t\\hline\n",
       "\t1 & 1.AIGNER\\_ZEB1\\_TARGETS & 0.995307 & $\\dots$ \\\\\n",
       "\t2 & 1.SANSOM\\_APC\\_TARGETS\\_UP & 0.856229 & $\\dots$ \\\\\n",
       "\t3 & 1.BCAT\\_GDS748\\_UP & 0.912969 & $\\dots$ \\\\\n",
       "\t4 & 1.BCAT.100\\_UP.V1\\_UP & 0.956058 & $\\dots$ \\\\\n",
       "\t5 & 1.PID\\_WNT\\_SIGNALING\\_PATHWAY & 0.393771 & $\\dots$ \\\\\n",
       "\t6 & 1.LIU\\_CDX2\\_TARGETS\\_UP & 0.938993 & $\\dots$ \\\\\n",
       "\t7 & 1.LUCAS\\_HNF4A\\_TARGETS\\_UP & 0.957338 & $\\dots$ \\\\\n",
       "\t8 & 10.LANG\\_MYB\\_FAMILY\\_TARGETS & 0.501956 & $\\dots$ \\\\\n",
       "\t9 & 10.PID\\_CMYB\\_PATHWAY & 0.717719 & $\\dots$ \\\\\n",
       "\t10 & 100.GO\\_CELL\\_CYCLE\\_G1\\_S\\_PHASE\\_TRANSITION & 0.797134 & $\\dots$ \\\\\n",
       "\t11 & 100.REACTOME\\_M\\_G1\\_TRANSITION & 0.839927 & $\\dots$ \\\\\n",
       "\t12 & 100.REACTOME\\_MITOTIC\\_M\\_M\\_G1\\_PHASES & 0.889786 & $\\dots$ \\\\\n",
       "\t13 & 100.KEGG\\_DNA\\_REPLICATION & 0.931974 & $\\dots$ \\\\\n",
       "\t14 & 100.REACTOME\\_DNA\\_REPLICATION & 0.880702 & $\\dots$ \\\\\n",
       "\t15 & 100.EGUCHI\\_CELL\\_CYCLE\\_RB1\\_TARGETS & 0.99677 & $\\dots$ \\\\\n",
       "\t16 & 100.WHITFIELD\\_CELL\\_CYCLE\\_LITERATURE & 0.986677 & $\\dots$ \\\\\n",
       "\t17 & 101.BIOCARTA\\_NFKB\\_PATHWAY & 0.866393 & $\\dots$ \\\\\n",
       "\t18 & 101.TIAN\\_TNF\\_SIGNALING\\_VIA\\_NFKB & 0.982528 & $\\dots$ \\\\\n",
       "\t19 & 101.BIOCARTA\\_TNFR2\\_PATHWAY & 0.994347 & $\\dots$ \\\\\n",
       "\t20 & 101.SANA\\_TNF\\_SIGNALING\\_UP & 0.873587 & $\\dots$ \\\\\n",
       "\t21 & 101.PHONG\\_TNF\\_TARGETS\\_UP & 0.913669 & $\\dots$ \\\\\n",
       "\t22 & 101.SCHOEN\\_NFKB\\_SIGNALING & 0.937307 & $\\dots$ \\\\\n",
       "\t23 & 101.GILMORE\\_CORE\\_NFKB\\_PATHWAY & 0.957348 & $\\dots$ \\\\\n",
       "\t24 & 102.REACTOME\\_ALPHA\\_LINOLENIC\\_ACID\\_ALA\\_METABOLISM & 0.991628 & $\\dots$ \\\\\n",
       "\t25 & 102.KEGG\\_BETA\\_ALANINE\\_METABOLISM & 0.912097 & $\\dots$ \\\\\n",
       "\t26 & 103.REACTOME\\_CELL\\_CYCLE & 0.877481 & $\\dots$ \\\\\n",
       "\t27 & 103.EGUCHI\\_CELL\\_CYCLE\\_RB1\\_TARGETS & 0.990458 & $\\dots$ \\\\\n",
       "\t28 & 103.REACTOME\\_REGULATION\\_OF\\_MITOTIC\\_CELL\\_CYCLE & 0.925191 & $\\dots$ \\\\\n",
       "\t29 & 104.ESW\\_FLI\\_UPREGULATED & 0.85987 & $\\dots$ \\\\\n",
       "\t30 & 105.REACTOME\\_THE\\_CITRIC\\_ACID\\_TCA\\_CYCLE\\_AND\\_RESPIRATORY\\_ELECTRON\\_TRANSPORT & 0.733066 & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m698×73 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0m│\u001b[1m Benchmark                         \u001b[0m\u001b[1m Is KS < extreme \u001b[0m\u001b[1m Is KS < area \u001b[0m\u001b[1m Is KS\u001b[0m ⋯\n",
       "\u001b[1m     \u001b[0m│\u001b[90m String                            \u001b[0m\u001b[90m Float64         \u001b[0m\u001b[90m Float64      \u001b[0m\u001b[90m Float\u001b[0m ⋯\n",
       "─────┼──────────────────────────────────────────────────────────────────────────\n",
       "   1 │ 1.AIGNER_ZEB1_TARGETS                     0.995307      0.997388        ⋯\n",
       "   2 │ 1.SANSOM_APC_TARGETS_UP                   0.856229      0.897693\n",
       "   3 │ 1.BCAT_GDS748_UP                          0.912969      0.954288\n",
       "   4 │ 1.BCAT.100_UP.V1_UP                       0.956058      0.96256\n",
       "   5 │ 1.PID_WNT_SIGNALING_PATHWAY               0.393771      0.21071         ⋯\n",
       "   6 │ 1.LIU_CDX2_TARGETS_UP                     0.938993      0.970396\n",
       "   7 │ 1.LUCAS_HNF4A_TARGETS_UP                  0.957338      0.964301\n",
       "   8 │ 10.LANG_MYB_FAMILY_TARGETS                0.501956      0.636833\n",
       "   9 │ 10.PID_CMYB_PATHWAY                       0.717719      0.774527        ⋯\n",
       "  10 │ 100.GO_CELL_CYCLE_G1_S_PHASE_TRA…         0.797134      0.889722\n",
       "  11 │ 100.REACTOME_M_G1_TRANSITION              0.839927      0.919116\n",
       "  ⋮  │                 ⋮                         ⋮              ⋮              ⋱\n",
       " 689 │ 97.CORDENONSI_YAP_CONSERVED_SIGN…         0.885126      0.939713\n",
       " 690 │ 97.HALLMARK_APICAL_JUNCTION               0.739366      0.817248        ⋯\n",
       " 691 │ 97.HALLMARK_EPITHELIAL_MESENCHYM…         0.847196      0.939443\n",
       " 692 │ 98.FOCAL_ADHESION                         0.907204      0.858194\n",
       " 693 │ 98.MISSIAGLIA_REGULATED_BY_METHY…         0.882784      0.92971\n",
       " 694 │ 99.KEGG_SPLICEOSOME                       0.949814      0.976611        ⋯\n",
       " 695 │ 99.KEGG_RNA_DEGRADATION                   0.874535      0.85537\n",
       " 696 │ 99.REACTOME_P53_DEPENDENT_G1_DNA…         0.925651      0.943198\n",
       " 697 │ 99.REACTOME_P53_INDEPENDENT_G1_S…         0.94842       0.96611\n",
       " 698 │ 99.REACTOME_REPAIR_SYNTHESIS_FOR…         0.953067      0.945585        ⋯\n",
       "\u001b[36m                                                 70 columns and 677 rows omitted\u001b[0m"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#\n",
    "benchmark_x_method = convert_vector_of_vector_to_matrix(benchmark_x_method_rows)\n",
    "\n",
    "benchmark_x_method = DataFrame(benchmark_x_method, method_)\n",
    "\n",
    "insertcols!(benchmark_x_method, 1, \"Benchmark\" => benchmark_)\n",
    "\n",
    "#\n",
    "benchmark_x_method_path = \"benchmark_x_method.tsv\"\n",
    "\n",
    "CSV.write(benchmark_x_method_path, benchmark_x_method)\n",
    "\n",
    "benchmark_x_method = read_data(benchmark_x_method_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8fc1e22d-36c3-40ea-bc53-bd60bba4a610",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "benchmark_x_method_metric = Matrix{Int64}(undef, length(benchmark_), n_method)\n",
    "\n",
    "for (benchmark_i, benchmark_score_) in\n",
    "    enumerate(eachrow(select(benchmark_x_method, Not(\"Benchmark\"))))\n",
    "\n",
    "    #\n",
    "    classic_score = benchmark_score_[\"Is KS < extreme\"]\n",
    "\n",
    "    #\n",
    "    benchmark_score_ = Vector(benchmark_score_)\n",
    "\n",
    "    #\n",
    "    if classic_score < 0\n",
    "\n",
    "        metric_ = benchmark_score_ .< classic_score\n",
    "\n",
    "    else\n",
    "\n",
    "        metric_ = classic_score .< benchmark_score_\n",
    "\n",
    "    end\n",
    "\n",
    "    #\n",
    "    benchmark_x_method_metric[benchmark_i, :] = metric_\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9391b22b-d86b-429e-92f6-7242b0838c98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sum_sort (generic function with 1 method)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function sum_sort(m::Matrix)::Nothing\n",
    "\n",
    "    for (method, n_better) in zip(sort_like(sum(eachrow(m)), method_)...)\n",
    "\n",
    "        println(method, \"    \", n_better)\n",
    "\n",
    "    end\n",
    "\n",
    "    return nothing\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef396782-c5ce-408f-9153-8859fe40dc04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======== < 1193 ========\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is JSD2 > extreme\n",
      "1    Is KS > area\n",
      "1    Is KS <> extreme\n",
      "1    Is JSD2 > area\n",
      "1    A KS > area\n",
      "1    A JSD1w > area\n",
      "1    A JSD2 > area\n",
      "1    A JSD2w > area\n",
      "2    Is JSD4 > area\n",
      "3    A KS > extreme\n",
      "3    A JSD2 > extreme\n",
      "3    A JSD4 > area\n",
      "4    A JSD1w > extreme\n",
      "4    A JSD2w > extreme\n",
      "5    Is JSD4 > extreme\n",
      "8    A JSD4 > extreme\n",
      "31    A JSD3 <> area\n",
      "38    A JSD3 <> extreme\n",
      "41    A JSD3 < area\n",
      "51    Is JSD3 <> extreme\n",
      "58    A JSD1 <> area\n",
      "67    A JSD3 < extreme\n",
      "69    A JSD4 <> extreme\n",
      "73    Is JSD3 <> area\n",
      "83    Is JSD3 < extreme\n",
      "86    Is JSD3 < area\n",
      "86    Is JSD4 <> extreme\n",
      "100    A JSD1 <> extreme\n",
      "106    A JSD1 > extreme\n",
      "111    Is JSD1 > extreme\n",
      "117    Is JSD3 > extreme\n",
      "121    A JSD3 > extreme\n",
      "131    Is JSD1 <> extreme\n",
      "139    Is JSD3 > area\n",
      "151    A JSD3 > area\n",
      "153    Is JSD1 <> area\n",
      "153    A JSD2 < extreme\n",
      "165    A JSD4 < extreme\n",
      "173    A JSD4 <> area\n",
      "175    A JSD1 < extreme\n",
      "187    Is JSD4 < extreme\n",
      "198    A KS < extreme\n",
      "198    A KS <> extreme\n",
      "209    Is JSD4 <> area\n",
      "225    A JSD1 < area\n",
      "261    A JSD1w <> extreme\n",
      "271    A JSD2w <> extreme\n",
      "292    Is JSD1 < extreme\n",
      "299    A JSD2 <> extreme\n",
      "341    A JSD1w < extreme\n",
      "346    A JSD2w < extreme\n",
      "403    Is JSD1 > area\n",
      "404    Is JSD1 < area\n",
      "413    Is JSD2 < extreme\n",
      "418    A JSD1 > area\n",
      "519    A JSD4 < area\n",
      "520    Is JSD2 < area\n",
      "521    A JSD2 < area\n",
      "523    Is JSD4 < area\n",
      "524    A KS < area\n",
      "524    A KS <> area\n",
      "526    Is KS < area\n",
      "526    Is KS <> area\n",
      "538    Is JSD2 <> area\n",
      "543    A JSD2 <> area\n",
      "561    Is JSD2 <> extreme\n",
      "566    A JSD1w < area\n",
      "569    A JSD2w < area\n",
      "571    A JSD1w <> area\n",
      "574    A JSD2w <> area\n",
      "======== 0 < 50 ========\n",
      "266\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD4 > area\n",
      "0    A KS > area\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > area\n",
      "1    A KS > extreme\n",
      "1    A JSD4 > area\n",
      "2    Is JSD4 > extreme\n",
      "2    A JSD2 > extreme\n",
      "2    A JSD4 > extreme\n",
      "3    A JSD1w > extreme\n",
      "3    A JSD2w > extreme\n",
      "4    A JSD3 < area\n",
      "6    A JSD3 <> area\n",
      "9    A JSD1 <> area\n",
      "11    Is JSD3 <> area\n",
      "16    Is JSD3 < area\n",
      "27    Is JSD1 <> area\n",
      "29    A JSD3 <> extreme\n",
      "39    Is JSD3 <> extreme\n",
      "48    A JSD1 < area\n",
      "48    A JSD3 < extreme\n",
      "48    A JSD4 <> extreme\n",
      "51    A JSD2 < extreme\n",
      "53    Is JSD3 < extreme\n",
      "55    A JSD4 <> area\n",
      "57    A JSD1 > extreme\n",
      "59    A JSD1 <> extreme\n",
      "61    Is JSD4 <> area\n",
      "62    Is JSD1 > extreme\n",
      "62    Is JSD4 <> extreme\n",
      "69    Is JSD3 > extreme\n",
      "70    A JSD3 > extreme\n",
      "75    Is JSD1 <> extreme\n",
      "77    A JSD3 > area\n",
      "78    Is JSD3 > area\n",
      "89    A JSD1 < extreme\n",
      "93    A KS < extreme\n",
      "93    A KS <> extreme\n",
      "105    Is JSD1 < area\n",
      "106    A JSD4 < extreme\n",
      "114    Is JSD4 < extreme\n",
      "120    Is JSD1 < extreme\n",
      "135    A JSD2 <> extreme\n",
      "135    A JSD2w <> extreme\n",
      "136    Is JSD2 < extreme\n",
      "137    A JSD1w <> extreme\n",
      "139    Is JSD1 > area\n",
      "140    A JSD1 > area\n",
      "157    Is JSD2 < area\n",
      "157    A JSD2 < area\n",
      "158    A JSD4 < area\n",
      "159    Is KS < area\n",
      "159    Is KS <> area\n",
      "159    A KS < area\n",
      "159    A KS <> area\n",
      "161    Is JSD4 < area\n",
      "164    Is JSD2 <> area\n",
      "166    A JSD1w < extreme\n",
      "166    A JSD2w < extreme\n",
      "167    A JSD2 <> area\n",
      "183    A JSD1w < area\n",
      "183    A JSD2w < area\n",
      "185    A JSD1w <> area\n",
      "185    A JSD2w <> area\n",
      "190    Is JSD2 <> extreme\n",
      "======== 50 < 100 ========\n",
      "130\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    A KS > area\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > area\n",
      "0    A JSD4 > area\n",
      "1    Is KS <> extreme\n",
      "1    A KS > extreme\n",
      "1    A JSD1w > extreme\n",
      "1    A JSD2 > extreme\n",
      "1    A JSD2w > extreme\n",
      "1    A JSD4 > extreme\n",
      "2    A JSD3 <> area\n",
      "3    A JSD3 <> extreme\n",
      "4    Is JSD3 <> extreme\n",
      "6    A JSD1 <> extreme\n",
      "6    A JSD3 < extreme\n",
      "6    A JSD3 < area\n",
      "7    Is JSD3 < extreme\n",
      "7    Is JSD3 <> area\n",
      "11    A JSD1 <> area\n",
      "12    Is JSD3 < area\n",
      "12    Is JSD4 <> extreme\n",
      "12    A JSD4 <> extreme\n",
      "13    A JSD4 < extreme\n",
      "14    Is JSD1 <> extreme\n",
      "17    Is JSD4 < extreme\n",
      "19    Is JSD1 > extreme\n",
      "22    Is JSD1 <> area\n",
      "24    A JSD1 > extreme\n",
      "25    Is JSD3 > area\n",
      "25    A JSD3 > extreme\n",
      "26    Is JSD3 > extreme\n",
      "27    A JSD1 < extreme\n",
      "29    A JSD3 > area\n",
      "31    A JSD4 <> area\n",
      "35    A KS < extreme\n",
      "35    A KS <> extreme\n",
      "36    Is JSD4 <> area\n",
      "38    A JSD1w <> extreme\n",
      "38    A JSD2 < extreme\n",
      "40    A JSD1 < area\n",
      "41    A JSD2w <> extreme\n",
      "48    Is JSD1 < extreme\n",
      "53    A JSD2 <> extreme\n",
      "56    A JSD1w < extreme\n",
      "56    A JSD2w < extreme\n",
      "73    Is JSD1 < area\n",
      "74    Is JSD2 < extreme\n",
      "81    Is JSD1 > area\n",
      "86    A JSD1 > area\n",
      "103    Is JSD2 < area\n",
      "103    A JSD2 < area\n",
      "104    Is KS < area\n",
      "104    Is KS <> area\n",
      "104    A KS < area\n",
      "104    A KS <> area\n",
      "106    Is JSD4 < area\n",
      "107    Is JSD2 <> extreme\n",
      "107    A JSD4 < area\n",
      "109    Is JSD2 <> area\n",
      "111    A JSD2 <> area\n",
      "112    A JSD1w < area\n",
      "113    A JSD2w < area\n",
      "116    A JSD1w <> area\n",
      "116    A JSD2w <> area\n",
      "======== 100 < 150 ========\n",
      "47\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD3 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD4 > area\n",
      "1    Is KS > area\n",
      "1    Is JSD2 > area\n",
      "1    Is JSD4 > extreme\n",
      "1    Is JSD4 > area\n",
      "1    Is JSD4 <> extreme\n",
      "1    A JSD3 <> area\n",
      "1    A JSD4 > extreme\n",
      "2    Is JSD3 < extreme\n",
      "3    Is JSD3 > extreme\n",
      "3    A JSD3 > extreme\n",
      "3    A JSD4 <> extreme\n",
      "5    Is JSD4 < extreme\n",
      "5    A JSD1 > extreme\n",
      "5    A JSD3 < area\n",
      "5    A JSD4 < extreme\n",
      "6    Is JSD1 > extreme\n",
      "6    A JSD1 < extreme\n",
      "6    A JSD1 <> extreme\n",
      "7    Is JSD1 <> extreme\n",
      "7    Is JSD3 > area\n",
      "7    A JSD1 <> area\n",
      "8    A JSD3 > area\n",
      "9    Is JSD3 < area\n",
      "9    Is JSD3 <> area\n",
      "9    A KS < extreme\n",
      "9    A KS <> extreme\n",
      "11    A JSD2 < extreme\n",
      "12    A JSD2 <> extreme\n",
      "13    A JSD1w <> extreme\n",
      "14    Is JSD1 < extreme\n",
      "15    A JSD2w <> extreme\n",
      "16    Is JSD1 <> area\n",
      "16    A JSD4 <> area\n",
      "17    Is JSD4 <> area\n",
      "17    A JSD1w < extreme\n",
      "17    A JSD2w < extreme\n",
      "18    A JSD1 < area\n",
      "28    A JSD1 > area\n",
      "31    Is JSD1 > area\n",
      "33    Is JSD1 < area\n",
      "33    Is JSD2 < extreme\n",
      "41    Is JSD4 < area\n",
      "41    A JSD4 < area\n",
      "42    Is KS < area\n",
      "42    Is KS <> area\n",
      "43    Is JSD2 <> extreme\n",
      "43    A KS < area\n",
      "43    A KS <> area\n",
      "43    A JSD1w < area\n",
      "43    A JSD2 < area\n",
      "43    A JSD2 <> area\n",
      "44    Is JSD2 < area\n",
      "44    Is JSD2 <> area\n",
      "44    A JSD1w <> area\n",
      "44    A JSD2w < area\n",
      "44    A JSD2w <> area\n",
      "======== 150 < 200 ========\n",
      "49\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    A KS > extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD3 > extreme\n",
      "1    Is JSD3 > area\n",
      "1    Is JSD4 > extreme\n",
      "1    Is JSD4 <> extreme\n",
      "1    A KS > area\n",
      "1    A JSD1 > extreme\n",
      "1    A JSD1w > area\n",
      "1    A JSD2 > area\n",
      "1    A JSD2w > area\n",
      "1    A JSD4 > area\n",
      "1    A JSD4 <> extreme\n",
      "2    A JSD3 <> extreme\n",
      "3    Is JSD1 > extreme\n",
      "3    Is JSD3 <> extreme\n",
      "3    A JSD3 > area\n",
      "3    A JSD4 > extreme\n",
      "4    A JSD3 < extreme\n",
      "5    Is JSD3 < extreme\n",
      "6    A JSD3 < area\n",
      "7    A JSD1 <> extreme\n",
      "8    A KS < extreme\n",
      "8    A KS <> extreme\n",
      "8    A JSD3 <> area\n",
      "8    A JSD4 < extreme\n",
      "9    A JSD1 <> area\n",
      "10    Is JSD1 <> extreme\n",
      "10    Is JSD4 < extreme\n",
      "10    A JSD2 < extreme\n",
      "11    Is JSD3 < area\n",
      "11    A JSD1 < extreme\n",
      "12    A JSD2 <> extreme\n",
      "13    Is JSD3 <> area\n",
      "13    A JSD1w <> extreme\n",
      "14    A JSD2w <> extreme\n",
      "16    A JSD4 <> area\n",
      "20    Is JSD1 > area\n",
      "20    Is JSD4 <> area\n",
      "20    A JSD1 > area\n",
      "20    A JSD1w < extreme\n",
      "21    Is JSD1 < extreme\n",
      "21    A JSD2w < extreme\n",
      "22    A JSD1 < area\n",
      "24    Is JSD1 <> area\n",
      "36    Is JSD1 < area\n",
      "36    Is JSD2 < extreme\n",
      "40    A JSD4 < area\n",
      "41    Is JSD4 < area\n",
      "41    A KS < area\n",
      "41    A KS <> area\n",
      "41    A JSD2 < area\n",
      "41    A JSD2 <> area\n",
      "42    Is KS < area\n",
      "42    Is KS <> area\n",
      "42    Is JSD2 < area\n",
      "42    Is JSD2 <> extreme\n",
      "42    Is JSD2 <> area\n",
      "42    A JSD1w < area\n",
      "42    A JSD2w < area\n",
      "43    A JSD1w <> area\n",
      "43    A JSD2w <> area\n",
      "======== 200 < 250 ========\n",
      "18\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 < area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD3 <> area\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1 <> area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 < area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 < extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    A JSD1 < extreme\n",
      "3    Is JSD1 > extreme\n",
      "3    Is JSD3 > extreme\n",
      "3    Is JSD3 > area\n",
      "3    A JSD1 > extreme\n",
      "3    A JSD3 > extreme\n",
      "3    A JSD3 > area\n",
      "4    Is JSD4 <> area\n",
      "4    A JSD4 <> area\n",
      "5    Is JSD1 <> area\n",
      "5    A JSD1w <> extreme\n",
      "5    A JSD2 < extreme\n",
      "5    A JSD2w <> extreme\n",
      "7    A KS < extreme\n",
      "7    A KS <> extreme\n",
      "7    A JSD1w < extreme\n",
      "7    A JSD2 <> extreme\n",
      "7    A JSD2w < extreme\n",
      "8    Is JSD1 < extreme\n",
      "9    A JSD1 < area\n",
      "12    Is JSD1 < area\n",
      "12    Is JSD1 > area\n",
      "12    Is JSD2 < extreme\n",
      "13    A JSD1 > area\n",
      "14    Is JSD4 < area\n",
      "14    A JSD4 < area\n",
      "15    Is JSD2 < area\n",
      "16    Is KS < area\n",
      "16    Is KS <> area\n",
      "16    Is JSD2 <> area\n",
      "16    A KS < area\n",
      "16    A KS <> area\n",
      "16    A JSD1w < area\n",
      "16    A JSD2 < area\n",
      "16    A JSD2w < area\n",
      "17    Is JSD2 <> extreme\n",
      "17    A JSD1w <> area\n",
      "17    A JSD2 <> area\n",
      "18    A JSD2w <> area\n",
      "======== 250 < 300 ========\n",
      "18\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 > extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD1w <> extreme\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD4 < extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD1 <> extreme\n",
      "1    Is JSD3 < extreme\n",
      "1    Is JSD3 > area\n",
      "1    Is JSD3 <> extreme\n",
      "1    A KS < extreme\n",
      "1    A KS <> extreme\n",
      "1    A JSD1 <> extreme\n",
      "1    A JSD2w <> extreme\n",
      "1    A JSD3 < extreme\n",
      "1    A JSD3 < area\n",
      "1    A JSD3 > extreme\n",
      "1    A JSD3 > area\n",
      "1    A JSD3 <> extreme\n",
      "2    A JSD1 < extreme\n",
      "2    A JSD1 <> area\n",
      "2    A JSD2 < extreme\n",
      "2    A JSD3 <> area\n",
      "3    A JSD1w < extreme\n",
      "4    Is JSD3 < area\n",
      "4    A JSD2w < extreme\n",
      "6    Is JSD1 < extreme\n",
      "6    Is JSD3 <> area\n",
      "6    A JSD2 <> extreme\n",
      "7    Is JSD1 <> area\n",
      "7    A JSD1 < area\n",
      "7    A JSD4 <> area\n",
      "8    Is JSD4 <> area\n",
      "12    A JSD1 > area\n",
      "14    Is JSD1 > area\n",
      "16    Is JSD1 < area\n",
      "16    Is JSD2 < area\n",
      "17    Is KS < area\n",
      "17    Is KS <> area\n",
      "17    Is JSD2 < extreme\n",
      "17    Is JSD2 <> area\n",
      "17    Is JSD4 < area\n",
      "17    A KS < area\n",
      "17    A KS <> area\n",
      "17    A JSD1w < area\n",
      "17    A JSD1w <> area\n",
      "17    A JSD2 < area\n",
      "17    A JSD2 <> area\n",
      "17    A JSD2w < area\n",
      "17    A JSD2w <> area\n",
      "17    A JSD4 < area\n",
      "18    Is JSD2 <> extreme\n",
      "======== 300 < 350 ========\n",
      "5\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 < extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD1 <> area\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 < area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD3 <> area\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 < extreme\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1 <> area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 < area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 < extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD3 > extreme\n",
      "1    Is JSD3 > area\n",
      "1    A JSD1 > extreme\n",
      "1    A JSD1w <> extreme\n",
      "1    A JSD3 > extreme\n",
      "2    A KS < extreme\n",
      "2    A KS <> extreme\n",
      "2    A JSD1 < area\n",
      "2    A JSD1w < extreme\n",
      "2    A JSD2 < extreme\n",
      "2    A JSD2w < extreme\n",
      "2    A JSD2w <> extreme\n",
      "2    A JSD3 > area\n",
      "3    Is JSD4 <> area\n",
      "3    A JSD2 <> extreme\n",
      "3    A JSD4 <> area\n",
      "5    Is KS < area\n",
      "5    Is KS <> area\n",
      "5    Is JSD1 < area\n",
      "5    Is JSD1 > area\n",
      "5    Is JSD2 < extreme\n",
      "5    Is JSD2 < area\n",
      "5    Is JSD2 <> extreme\n",
      "5    Is JSD2 <> area\n",
      "5    Is JSD4 < area\n",
      "5    A KS < area\n",
      "5    A KS <> area\n",
      "5    A JSD1 > area\n",
      "5    A JSD1w < area\n",
      "5    A JSD1w <> area\n",
      "5    A JSD2 < area\n",
      "5    A JSD2 <> area\n",
      "5    A JSD2w < area\n",
      "5    A JSD2w <> area\n",
      "5    A JSD4 < area\n",
      "======== 350 < 400 ========\n",
      "8\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD1 <> extreme\n",
      "1    Is JSD1 <> area\n",
      "1    Is JSD3 > area\n",
      "1    Is JSD3 <> area\n",
      "1    A JSD1 < extreme\n",
      "1    A JSD1 > extreme\n",
      "1    A JSD1 <> area\n",
      "1    A JSD3 < area\n",
      "1    A JSD3 > area\n",
      "1    A JSD4 < extreme\n",
      "2    Is JSD1 < extreme\n",
      "2    Is JSD3 < area\n",
      "2    A KS < extreme\n",
      "2    A KS <> extreme\n",
      "2    A JSD2 < extreme\n",
      "3    Is JSD4 <> area\n",
      "3    A JSD1w <> extreme\n",
      "3    A JSD2w <> extreme\n",
      "3    A JSD4 <> area\n",
      "4    Is JSD2 < extreme\n",
      "4    A JSD1w < extreme\n",
      "4    A JSD2 <> extreme\n",
      "4    A JSD2w < extreme\n",
      "5    A JSD1 < area\n",
      "6    A JSD1 > area\n",
      "7    Is JSD1 > area\n",
      "8    Is KS < area\n",
      "8    Is KS <> area\n",
      "8    Is JSD1 < area\n",
      "8    Is JSD2 < area\n",
      "8    Is JSD2 <> extreme\n",
      "8    Is JSD2 <> area\n",
      "8    Is JSD4 < area\n",
      "8    A KS < area\n",
      "8    A KS <> area\n",
      "8    A JSD1w < area\n",
      "8    A JSD1w <> area\n",
      "8    A JSD2 < area\n",
      "8    A JSD2 <> area\n",
      "8    A JSD2w < area\n",
      "8    A JSD2w <> area\n",
      "8    A JSD4 < area\n",
      "======== 400 < 450 ========\n",
      "5\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD1 <> area\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 < area\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 > area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD3 <> area\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    Is JSD4 <> area\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 < extreme\n",
      "0    A JSD1 > extreme\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1 <> area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD1w <> extreme\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 < area\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 > area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 < extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "0    A JSD4 <> area\n",
      "1    Is JSD1 > extreme\n",
      "1    A KS < extreme\n",
      "1    A KS <> extreme\n",
      "1    A JSD1 < area\n",
      "1    A JSD1w < extreme\n",
      "1    A JSD2 < extreme\n",
      "2    Is JSD1 < extreme\n",
      "2    A JSD2 <> extreme\n",
      "2    A JSD2w < extreme\n",
      "2    A JSD2w <> extreme\n",
      "3    Is JSD1 > area\n",
      "3    Is JSD2 < extreme\n",
      "3    A JSD1 > area\n",
      "4    Is JSD1 < area\n",
      "4    Is JSD2 <> extreme\n",
      "4    Is JSD4 < area\n",
      "4    A KS < area\n",
      "4    A KS <> area\n",
      "4    A JSD1w < area\n",
      "4    A JSD2 < area\n",
      "4    A JSD2 <> area\n",
      "4    A JSD4 < area\n",
      "5    Is KS < area\n",
      "5    Is KS <> area\n",
      "5    Is JSD2 < area\n",
      "5    Is JSD2 <> area\n",
      "5    A JSD1w <> area\n",
      "5    A JSD2w < area\n",
      "5    A JSD2w <> area\n",
      "======== 450 < 500 ========\n",
      "7\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 <> area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 < area\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD1 <> extreme\n",
      "1    Is JSD3 > area\n",
      "1    Is JSD3 <> area\n",
      "1    Is JSD4 < extreme\n",
      "1    A JSD1 > extreme\n",
      "1    A JSD1 <> extreme\n",
      "1    A JSD2 < extreme\n",
      "1    A JSD3 > area\n",
      "2    A KS < extreme\n",
      "2    A KS <> extreme\n",
      "2    A JSD1 < extreme\n",
      "2    A JSD1w <> extreme\n",
      "2    A JSD2 <> extreme\n",
      "2    A JSD2w <> extreme\n",
      "2    A JSD4 < extreme\n",
      "3    Is JSD3 < area\n",
      "3    A JSD1w < extreme\n",
      "3    A JSD4 <> area\n",
      "4    Is JSD4 <> area\n",
      "4    A JSD2w < extreme\n",
      "5    Is JSD1 < extreme\n",
      "5    Is JSD1 <> area\n",
      "5    A JSD1 < area\n",
      "6    Is JSD1 < area\n",
      "6    Is JSD1 > area\n",
      "6    Is JSD2 < extreme\n",
      "6    Is JSD2 <> extreme\n",
      "6    A JSD1 > area\n",
      "7    Is KS < area\n",
      "7    Is KS <> area\n",
      "7    Is JSD2 < area\n",
      "7    Is JSD2 <> area\n",
      "7    Is JSD4 < area\n",
      "7    A KS < area\n",
      "7    A KS <> area\n",
      "7    A JSD1w < area\n",
      "7    A JSD1w <> area\n",
      "7    A JSD2 < area\n",
      "7    A JSD2 <> area\n",
      "7    A JSD2w < area\n",
      "7    A JSD2w <> area\n",
      "7    A JSD4 < area\n",
      "======== 500 < 550 ========\n",
      "2\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 < extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD1 <> area\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 < area\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 > area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD3 <> area\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 > extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 > area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD1 < area\n",
      "1    Is JSD1 > area\n",
      "1    Is JSD2 < extreme\n",
      "1    Is JSD4 < extreme\n",
      "1    Is JSD4 <> area\n",
      "1    A KS < extreme\n",
      "1    A KS <> extreme\n",
      "1    A JSD1 < extreme\n",
      "1    A JSD1 < area\n",
      "1    A JSD1 <> extreme\n",
      "1    A JSD1 <> area\n",
      "1    A JSD1w < extreme\n",
      "1    A JSD1w <> extreme\n",
      "1    A JSD2 < extreme\n",
      "1    A JSD2 <> extreme\n",
      "1    A JSD2w < extreme\n",
      "1    A JSD2w <> extreme\n",
      "1    A JSD3 < area\n",
      "1    A JSD3 <> area\n",
      "1    A JSD4 < extreme\n",
      "1    A JSD4 <> area\n",
      "2    Is KS < area\n",
      "2    Is KS <> area\n",
      "2    Is JSD2 < area\n",
      "2    Is JSD2 <> extreme\n",
      "2    Is JSD2 <> area\n",
      "2    Is JSD4 < area\n",
      "2    A KS < area\n",
      "2    A KS <> area\n",
      "2    A JSD1 > area\n",
      "2    A JSD1w < area\n",
      "2    A JSD1w <> area\n",
      "2    A JSD2 < area\n",
      "2    A JSD2 <> area\n",
      "2    A JSD2w < area\n",
      "2    A JSD2w <> area\n",
      "2    A JSD4 < area\n",
      "======== 550 < 600 ========\n",
      "23\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 < area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD3 <> area\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "1    Is JSD4 <> extreme\n",
      "1    A JSD1 < extreme\n",
      "1    A JSD1 <> area\n",
      "1    A JSD4 < extreme\n",
      "1    A JSD4 <> extreme\n",
      "2    Is JSD1 > extreme\n",
      "2    Is JSD1 <> extreme\n",
      "2    Is JSD3 < area\n",
      "2    Is JSD3 <> area\n",
      "2    Is JSD4 < extreme\n",
      "3    A KS < extreme\n",
      "3    A KS <> extreme\n",
      "3    A JSD1 > extreme\n",
      "3    A JSD2 < extreme\n",
      "4    Is JSD3 > extreme\n",
      "4    Is JSD3 > area\n",
      "4    A JSD1 < area\n",
      "4    A JSD3 > extreme\n",
      "4    A JSD3 > area\n",
      "4    A JSD4 <> area\n",
      "5    Is JSD1 <> area\n",
      "5    A JSD1w < extreme\n",
      "5    A JSD1w <> extreme\n",
      "5    A JSD2w <> extreme\n",
      "6    Is JSD1 < extreme\n",
      "6    A JSD2w < extreme\n",
      "7    A JSD2 <> extreme\n",
      "9    Is JSD4 <> area\n",
      "12    Is JSD1 > area\n",
      "13    Is JSD1 < area\n",
      "15    A JSD1 > area\n",
      "18    Is JSD2 < extreme\n",
      "18    A JSD4 < area\n",
      "19    Is JSD4 < area\n",
      "21    Is KS < area\n",
      "21    Is KS <> area\n",
      "21    Is JSD2 < area\n",
      "21    Is JSD2 <> area\n",
      "21    A KS < area\n",
      "21    A KS <> area\n",
      "21    A JSD1w < area\n",
      "21    A JSD2 < area\n",
      "21    A JSD2 <> area\n",
      "21    A JSD2w < area\n",
      "22    Is JSD2 <> extreme\n",
      "22    A JSD1w <> area\n",
      "22    A JSD2w <> area\n",
      "======== 600 < 650 ========\n",
      "1\n",
      "======== 650 < 700 ========\n",
      "4\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 > area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 > extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 > area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD3 < area\n",
      "1    Is JSD3 <> area\n",
      "1    Is JSD4 < extreme\n",
      "1    A KS < extreme\n",
      "1    A KS <> extreme\n",
      "1    A JSD1 < extreme\n",
      "1    A JSD1 <> extreme\n",
      "1    A JSD1w <> extreme\n",
      "1    A JSD2 < extreme\n",
      "1    A JSD2 <> extreme\n",
      "1    A JSD2w <> extreme\n",
      "1    A JSD3 <> area\n",
      "1    A JSD4 < extreme\n",
      "2    Is JSD1 < extreme\n",
      "2    Is JSD1 <> area\n",
      "2    Is JSD2 < extreme\n",
      "2    A JSD1 <> area\n",
      "2    A JSD1w < extreme\n",
      "2    A JSD2w < extreme\n",
      "2    A JSD3 < area\n",
      "2    A JSD4 <> area\n",
      "3    Is JSD1 < area\n",
      "3    Is JSD1 > area\n",
      "3    Is JSD4 <> area\n",
      "4    Is KS < area\n",
      "4    Is KS <> area\n",
      "4    Is JSD2 < area\n",
      "4    Is JSD2 <> extreme\n",
      "4    Is JSD2 <> area\n",
      "4    Is JSD4 < area\n",
      "4    A KS < area\n",
      "4    A KS <> area\n",
      "4    A JSD1 < area\n",
      "4    A JSD1 > area\n",
      "4    A JSD1w < area\n",
      "4    A JSD1w <> area\n",
      "4    A JSD2 < area\n",
      "4    A JSD2 <> area\n",
      "4    A JSD2w < area\n",
      "4    A JSD2w <> area\n",
      "4    A JSD4 < area\n",
      "======== 700 < 750 ========\n",
      "0\n",
      "======== 750 < 800 ========\n",
      "3\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD1 > extreme\n",
      "0    Is JSD1 <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD3 < extreme\n",
      "0    Is JSD3 > extreme\n",
      "0    Is JSD3 > area\n",
      "0    Is JSD3 <> extreme\n",
      "0    Is JSD4 < extreme\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    Is JSD4 <> extreme\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1 < extreme\n",
      "0    A JSD1 > extreme\n",
      "0    A JSD1 <> extreme\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD3 < extreme\n",
      "0    A JSD3 > extreme\n",
      "0    A JSD3 > area\n",
      "0    A JSD3 <> extreme\n",
      "0    A JSD4 < extreme\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "0    A JSD4 <> extreme\n",
      "1    Is JSD1 < extreme\n",
      "1    Is JSD1 <> area\n",
      "1    Is JSD3 < area\n",
      "1    Is JSD3 <> area\n",
      "1    A KS < extreme\n",
      "1    A KS <> extreme\n",
      "1    A JSD1 <> area\n",
      "1    A JSD1w < extreme\n",
      "1    A JSD1w <> extreme\n",
      "1    A JSD2 < extreme\n",
      "1    A JSD2 <> extreme\n",
      "1    A JSD2w < extreme\n",
      "1    A JSD2w <> extreme\n",
      "1    A JSD3 < area\n",
      "1    A JSD3 <> area\n",
      "1    A JSD4 <> area\n",
      "2    Is JSD1 > area\n",
      "2    Is JSD2 < extreme\n",
      "2    Is JSD4 <> area\n",
      "2    A JSD1 < area\n",
      "2    A JSD1 > area\n",
      "3    Is KS < area\n",
      "3    Is KS <> area\n",
      "3    Is JSD1 < area\n",
      "3    Is JSD2 < area\n",
      "3    Is JSD2 <> extreme\n",
      "3    Is JSD2 <> area\n",
      "3    Is JSD4 < area\n",
      "3    A KS < area\n",
      "3    A KS <> area\n",
      "3    A JSD1w < area\n",
      "3    A JSD1w <> area\n",
      "3    A JSD2 < area\n",
      "3    A JSD2 <> area\n",
      "3    A JSD2w < area\n",
      "3    A JSD2w <> area\n",
      "3    A JSD4 < area\n",
      "======== 800 < 850 ========\n",
      "37\n",
      "0    Is KS < extreme\n",
      "0    Is KS > extreme\n",
      "0    Is KS > area\n",
      "0    Is KS <> extreme\n",
      "0    Is JSD2 > extreme\n",
      "0    Is JSD2 > area\n",
      "0    Is JSD4 > extreme\n",
      "0    Is JSD4 > area\n",
      "0    A KS > extreme\n",
      "0    A KS > area\n",
      "0    A JSD1w > extreme\n",
      "0    A JSD1w > area\n",
      "0    A JSD2 > extreme\n",
      "0    A JSD2 > area\n",
      "0    A JSD2w > extreme\n",
      "0    A JSD2w > area\n",
      "0    A JSD4 > extreme\n",
      "0    A JSD4 > area\n",
      "2    A JSD3 <> extreme\n",
      "3    Is JSD3 <> extreme\n",
      "3    A JSD3 <> area\n",
      "3    A JSD4 <> extreme\n",
      "4    A JSD3 < area\n",
      "5    Is JSD1 > extreme\n",
      "5    Is JSD4 <> extreme\n",
      "5    A JSD1 <> area\n",
      "6    A JSD1 > extreme\n",
      "7    Is JSD3 > extreme\n",
      "7    Is JSD3 <> area\n",
      "7    A JSD3 < extreme\n",
      "8    Is JSD3 < area\n",
      "8    A JSD2 < extreme\n",
      "9    A JSD4 <> area\n",
      "10    A JSD3 > extreme\n",
      "11    Is JSD1 <> extreme\n",
      "11    Is JSD3 < extreme\n",
      "11    Is JSD3 > area\n",
      "11    Is JSD4 <> area\n",
      "12    Is JSD1 <> area\n",
      "12    A JSD1 <> extreme\n",
      "13    Is JSD2 < extreme\n",
      "13    A JSD3 > area\n",
      "15    A KS < extreme\n",
      "15    A KS <> extreme\n",
      "18    A JSD1 < area\n",
      "21    Is JSD1 > area\n",
      "21    A JSD1 < extreme\n",
      "22    A JSD4 < extreme\n",
      "24    Is JSD4 < extreme\n",
      "24    A JSD1 > area\n",
      "26    Is JSD2 < area\n",
      "27    A KS < area\n",
      "27    A KS <> area\n",
      "27    A JSD1w <> extreme\n",
      "27    A JSD2 < area\n",
      "27    A JSD2w <> extreme\n",
      "28    Is KS < area\n",
      "28    Is KS <> area\n",
      "28    Is JSD1 < extreme\n",
      "28    Is JSD2 <> extreme\n",
      "28    Is JSD2 <> area\n",
      "28    A JSD2 <> extreme\n",
      "29    Is JSD1 < area\n",
      "29    A JSD2 <> area\n",
      "31    A JSD2w < extreme\n",
      "31    A JSD4 < area\n",
      "32    Is JSD4 < area\n",
      "32    A JSD1w < extreme\n",
      "33    A JSD1w < area\n",
      "33    A JSD1w <> area\n",
      "33    A JSD2w < area\n",
      "33    A JSD2w <> area\n",
      "======== 850 < 900 ========\n",
      "0\n",
      "======== 900 < 950 ========\n",
      "0\n",
      "======== 950 < 1000 ========\n",
      "1\n",
      "======== 1000 < 1050 ========\n",
      "1\n",
      "======== 1050 < 1100 ========\n",
      "0\n",
      "======== 1100 < 1150 ========\n",
      "1\n",
      "======== 1150 < 1200 ========\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "buffer = \"=\"^8\n",
    "\n",
    "benchmark_size_maximum = maximum(benchmark_size_)\n",
    "\n",
    "println(buffer, \" < \", benchmark_size_maximum, \" \", buffer)\n",
    "\n",
    "sum_sort(benchmark_x_method_metric)\n",
    "\n",
    "increment = 50\n",
    "\n",
    "for (minimum_size, maximum_size) in [\n",
    "    ((i - 1) * increment, i * increment) for\n",
    "    i = 1:Int64(ceil(benchmark_size_maximum / increment))\n",
    "]\n",
    "\n",
    "    #\n",
    "    println(buffer, \" \", minimum_size, \" < \", maximum_size, \" \", buffer)\n",
    "\n",
    "    #\n",
    "    is_selected = minimum_size .< benchmark_size_ .< maximum_size\n",
    "    \n",
    "    n_selected = sum(is_selected)\n",
    "    \n",
    "    println(n_selected)\n",
    "\n",
    "    if 1 < n_selected\n",
    "\n",
    "        sum_sort(benchmark_x_method_metric[is_selected, :])\n",
    "\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "809a8c1f-66bc-46d0-8a6d-c5e1981f68c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.1",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
