{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76cde750-8ccd-434f-82f8-94e84d6a6b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "using BenchmarkTools\n",
    "using Revise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "983f0dc7-21c5-4531-a5c2-6790849cf78d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling FeatureSetEnrichment [d7bef1b9-4643-4ed1-9ceb-ed43dd382ec0]\n",
      "└ @ Base loading.jl:1317\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<script>\n",
       "// Immediately-invoked-function-expression to avoid global variables.\n",
       "(function() {\n",
       "    var warning_div = document.getElementById(\"webio-warning-15656338930442271228\");\n",
       "    var hide = function () {\n",
       "        var script = document.getElementById(\"webio-setup-15281551830535315427\");\n",
       "        var parent = script && script.parentElement;\n",
       "        var grandparent = parent && parent.parentElement;\n",
       "        if (grandparent) {\n",
       "            grandparent.style.display = \"none\";\n",
       "        }\n",
       "        warning_div.style.display = \"none\";\n",
       "    };\n",
       "    if (typeof Jupyter !== \"undefined\") {\n",
       "        console.log(\"WebIO detected Jupyter notebook environment.\");\n",
       "        // Jupyter notebook.\n",
       "        var extensions = (\n",
       "            Jupyter\n",
       "            && Jupyter.notebook.config.data\n",
       "            && Jupyter.notebook.config.data.load_extensions\n",
       "        );\n",
       "        if (extensions && extensions[\"webio-jupyter-notebook\"]) {\n",
       "            // Extension already loaded.\n",
       "            console.log(\"Jupyter WebIO nbextension detected; not loading ad-hoc.\");\n",
       "            hide();\n",
       "            return;\n",
       "        }\n",
       "    } else if (window.location.pathname.includes(\"/lab\")) {\n",
       "        // Guessing JupyterLa\n",
       "        console.log(\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\");\n",
       "        hide();\n",
       "        return;\n",
       "    }\n",
       "})();\n",
       "\n",
       "</script>\n",
       "<p\n",
       "    id=\"webio-warning-15656338930442271228\"\n",
       "    class=\"output_text output_stderr\"\n",
       "    style=\"padding: 1em; font-weight: bold;\"\n",
       ">\n",
       "    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\n",
       "    For troubleshooting, please see <a href=\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\">\n",
       "    the WebIO/IJulia documentation</a>.\n",
       "    <!-- TODO: link to installation docs. -->\n",
       "</p>\n"
      ],
      "text/plain": [
       "HTML{String}(\"<script>\\n// Immediately-invoked-function-expression to avoid global variables.\\n(function() {\\n    var warning_div = document.getElementById(\\\"webio-warning-15656338930442271228\\\");\\n    var hide = function () {\\n        var script = document.getElementById(\\\"webio-setup-15281551830535315427\\\");\\n        var parent = script && script.parentElement;\\n        var grandparent = parent && parent.parentElement;\\n        if (grandparent) {\\n            grandparent.style.display = \\\"none\\\";\\n        }\\n        warning_div.style.display = \\\"none\\\";\\n    };\\n    if (typeof Jupyter !== \\\"undefined\\\") {\\n        console.log(\\\"WebIO detected Jupyter notebook environment.\\\");\\n        // Jupyter notebook.\\n        var extensions = (\\n            Jupyter\\n            && Jupyter.notebook.config.data\\n            && Jupyter.notebook.config.data.load_extensions\\n        );\\n        if (extensions && extensions[\\\"webio-jupyter-notebook\\\"]) {\\n            // Extension already loaded.\\n            console.log(\\\"Jupyter WebIO nbextension detected; not loading ad-hoc.\\\");\\n            hide();\\n            return;\\n        }\\n    } else if (window.location.pathname.includes(\\\"/lab\\\")) {\\n        // Guessing JupyterLa\\n        console.log(\\\"Jupyter Lab detected; make sure the @webio/jupyter-lab-provider labextension is installed.\\\");\\n        hide();\\n        return;\\n    }\\n})();\\n\\n</script>\\n<p\\n    id=\\\"webio-warning-15656338930442271228\\\"\\n    class=\\\"output_text output_stderr\\\"\\n    style=\\\"padding: 1em; font-weight: bold;\\\"\\n>\\n    Unable to load WebIO. Please make sure WebIO works for your Jupyter client.\\n    For troubleshooting, please see <a href=\\\"https://juliagizmos.github.io/WebIO.jl/latest/providers/ijulia/\\\">\\n    the WebIO/IJulia documentation</a>.\\n    <!-- TODO: link to installation docs. -->\\n</p>\\n\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "using CSV\n",
    "using DataFrames\n",
    "using JSON\n",
    "\n",
    "using DataIO\n",
    "using GCTGMT\n",
    "using Normalization\n",
    "using Support\n",
    "\n",
    "using FeatureSetEnrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f25618fe-79dc-49e0-bee4-6ac273975ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "name_to_set_to_element_ = Dict(splitpath(path)[end] => read_gmt(path) for path in read_directory(\"/Users/kwat/Downloads/gene_sets/\"))\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24b75e0a-3c53-4fd9-aa33-71edd712f6db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n",
      "The minimum and the maximum have the same absolute value; returning the minimum.\n"
     ]
    }
   ],
   "source": [
    "method_ = Vector{String}()\n",
    "\n",
    "for method in keys(score_set_new([\"a\", \"b\"], [-1.0, 1.0], [\"a\"]; plot = false))\n",
    "\n",
    "    push!(method_, string(method, \" extreme\"), string(method, \" area\"))\n",
    "\n",
    "end\n",
    "\n",
    "n_method = length(method_)\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "028f61ac-6359-49db-94c4-cfd07db70a26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/Users/kwat/Downloads/benchmarks/\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmarks_directory_path = \"/Users/kwat/Downloads/benchmarks/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25f54201-7b3d-4edf-93cd-ce9b5415b07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_ = Vector{String}()\n",
    "\n",
    "benchmark_x_method_rows = Vector{Vector{Float64}}()\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d0821e8-508d-40ad-84ad-c53b37dc3d84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99\n",
      "    4762 set.\n",
      "    KEGG_SPLICEOSOME\n",
      "    KEGG_RNA_DEGRADATION\n",
      "    REACTOME_P53_DEPENDENT_G1_DNA_DAMAGE_RESPONSE\n",
      "    REACTOME_P53_INDEPENDENT_G1_S_DNA_DAMAGE_CHECKPOINT\n",
      "    REACTOME_REPAIR_SYNTHESIS_FOR_GAP_FILLING_BY_DNA_POL_IN_TC_NER\n",
      "98\n",
      "    4811 set.\n",
      "    FOCAL_ADHESION\n",
      "    LIANG_SILENCED_BY_METHYLATION (missing)\n",
      "    MISSIAGLIA_REGULATED_BY_METHYLATION_UP\n",
      "97\n",
      "    5001 set.\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE\n",
      "    HALLMARK_APICAL_JUNCTION\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION\n",
      "96\n",
      "    14765 set.\n",
      "    GO_DNA_REPLICATION\n",
      "    GO_DNA_DEPENDENT_DNA_REPLICATION\n",
      "    GO_REGULATION_OF_DNA_DEPENDENT_DNA_REPLICATION\n",
      "    GO_DNA_STRAND_ELONGATION\n",
      "    GO_DNA_STRAND_ELONGATION_INVOLVED_IN_DNA_REPLICATION\n",
      "95\n",
      "    6111 set.\n",
      "    AML_TCP_vs_untreated_UP\n",
      "94\n",
      "    4811 set.\n",
      "    Erythroid_associated_genes\n",
      "93\n",
      "    4811 set.\n",
      "    Amino_Acid_Metabolism\n",
      "92\n",
      "    5917 set.\n",
      "    GO_REGULATION_OF_MONONUCLEAR_CELL_MIGRATION\n",
      "91\n",
      "    4811 set.\n",
      "    LSC_ASSOCIATED_GENES\n",
      "90\n",
      "    4762 set.\n",
      "    SCHUHMACHER_MYC_TARGETS_UP\n",
      "    DANG_MYC_TARGETS_UP\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS (missing)\n",
      "    CAIRO_PML_TARGETS_BOUND_BY_MYC_UP\n",
      "9\n",
      "    4812 set.\n",
      "    HINATA_NFKB_TARGETS_FIBROBLAST_UP\n",
      "    SCHOEN_NFKB_SIGNALING\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
      "    TIAN_TNF_SIGNALING_VIA_NFKB\n",
      "89\n",
      "    20994 set.\n",
      "    KEGG_HOMOLOGOUS_RECOMBINATION\n",
      "    GO_MITOTIC_NUCLEAR_DIVISION\n",
      "    GO_REGULATION_OF_TRANSCRIPTION_INVOLVED_IN_G1_S_TRANSITION_OF_MITOTIC_CELL_CYCLE\n",
      "88\n",
      "    4762 set.\n",
      "    SCHLOSSER_MYC_AND_SERUM_RESPONSE_SYNERGY\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP\n",
      "    COLLER_MYC_TARGETS_UP\n",
      "    SCHUHMACHER_MYC_TARGETS_UP\n",
      "    BILANGES_SERUM_AND_RAPAMYCIN_SENSITIVE_GENES\n",
      "87\n",
      "    6328 set.\n",
      "    activated_treg\n",
      "    HALLMARK_IL2_STAT5_SIGNALING\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "    HALLMARK_E2F_TARGETS\n",
      "86\n",
      "    4762 set.\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING\n",
      "    ZHANG_INTERFERON_RESPONSE\n",
      "    EINAV_INTERFERON_SIGNATURE_IN_CANCER\n",
      "    ST_INTERFERON_GAMMA_PATHWAY\n",
      "85\n",
      "    4762 set.\n",
      "    WNT_SIGNALING\n",
      "    YU_MYC_TARGETS_UP\n",
      "    KEGG_CELL_CYCLE\n",
      "    REACTOME_CELL_CYCLE\n",
      "84\n",
      "    4762 set.\n",
      "    ZHU_CMV_8_HR_UP\n",
      "    DER_IFN_ALPHA_RESPONSE_UP\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING\n",
      "    EINAV_INTERFERON_SIGNATURE_IN_CANCER\n",
      "    BROWNE_INTERFERON_RESPONSIVE_GENES\n",
      "83\n",
      "    4762 set.\n",
      "    DANG_REGULATED_BY_MYC_UP\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP\n",
      "    SCHUHMACHER_MYC_TARGETS (missing)\n",
      "    COLLER_MYC_TARGETS (missing)\n",
      "    YU_MYC_TARGETS (missing)\n",
      "    DANG_REGULATED_BY_MYC (missing)\n",
      "82\n",
      "    4812 set.\n",
      "    BILD_MYC_ONCOGENIC_SIGNATURE\n",
      "    SCHUHMACHER_MYC_TARGETS_UP\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    HALLMARK_MYC_TARGETS_V2\n",
      "    COLLER_MYC_TARGETS_UP\n",
      "81\n",
      "    6229 set.\n",
      "    MILI_PSEUDOPODIA_HAPTOTAXIS_UP\n",
      "    HAMAI_APOPTOSIS_VIA_TRAIL_UP\n",
      "    SEIDEN_ONCOGENESIS_BY_MET\n",
      "80\n",
      "    239 set.\n",
      "    HALLMARK_KRAS_SIGNALING_UP\n",
      "    EGFR_UP.V1 (missing)\n",
      "    KRAS.LUNG.BREAST_UP.V1 (missing)\n",
      "8\n",
      "    4762 set.\n",
      "    VANTVEER_BREAST_CANCER_ESR1 (missing)\n",
      "    DOANE_BREAST_CANCER_ESR1 (missing)\n",
      "79\n",
      "    4762 set.\n",
      "    KEGG_LYSOSOME\n",
      "    KEGG_REGULATION_OF_AUTOPHAGY\n",
      "    KEGG_FATTY_ACID_METABOLISM\n",
      "    KEGG_PEROXISOME\n",
      "    REACTOME_RESPIRATORY_ELECTRON_TRANSPORT\n",
      "78\n",
      "    4762 set.\n",
      "    KEGG_STEROID_BIOSYNTHESIS\n",
      "    KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
      "    KEGG_PEROXISOME\n",
      "77\n",
      "    4762 set.\n",
      "    KONG_E2F3_TARGETS\n",
      "    REACTOME_CYCLIN_A_B1_ASSOCIATED_EVENTS_DURING_G2_M_TRANSITION\n",
      "    WHITFIELD_CELL_CYCLE_G2\n",
      "    REACTOME_GENERIC_TRANSCRIPTION_PATHWAY\n",
      "76\n",
      "    4762 set.\n",
      "    ZHANG_TARGETS_OF_EWSR1_FLI1_FUSION\n",
      "    RIGGI_EWING_SARCOMA_PROGENITOR_UP\n",
      "    STAEGE_EWING_FAMILY_TUMOR\n",
      "75\n",
      "    5648 set.\n",
      "    SCHUHMACHER_MYC_TARGETS_UP\n",
      "    HALLMARK_MYC_TARGETS_V2\n",
      "    MENSSEN_MYC_TARGETS\n",
      "    SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP\n",
      "74\n",
      "    5917 set.\n",
      "    GO_ANTIGEN_BINDING\n",
      "    GO_ALPHA_BETA_T_CELL_ACTIVATION\n",
      "73\n",
      "    4762 set.\n",
      "    PENG_RAPAMYCIN_RESPONSE_DN\n",
      "72\n",
      "    50 set.\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "71\n",
      "    6229 set.\n",
      "    LEE_BMP2_TARGETS_UP\n",
      "    NABA_MATRISOME\n",
      "    NABA_MATRISOME_ASSOCIATED\n",
      "    NABA_CORE_MATRISOME\n",
      "    KEGG_LYSOSOME\n",
      "70\n",
      "    4762 set.\n",
      "    KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "    BIOCARTA_GLYCOLYSIS_PATHWAY\n",
      "    REACTOME_GLYCOLYSIS\n",
      "    MOOTHA_GLYCOLYSIS\n",
      "7\n",
      "    4762 set.\n",
      "    KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
      "    KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
      "69\n",
      "    6229 set.\n",
      "    PUJANA_BREAST_CANCER_WITH_BRCA1_MUTATED_UP\n",
      "    COLLER_MYC_TARGETS_UP\n",
      "68\n",
      "    4762 set.\n",
      "    REACTOME_TRANSCRIPTION\n",
      "    REACTOME_TRANSLATION\n",
      "    KEGG_DNA_REPLICATION\n",
      "67\n",
      "    4762 set.\n",
      "    REACTOME_DNA_REPLICATION\n",
      "    KEGG_DNA_REPLICATION\n",
      "66\n",
      "    14765 set.\n",
      "    GO_T_CELL_DIFFERENTIATION\n",
      "    GO_REGULATION_OF_LEUKOCYTE_CHEMOTAXIS\n",
      "    GO_LEUKOCYTE_CELL_CELL_ADHESION\n",
      "    GO_ADAPTIVE_IMMUNE_RESPONSE\n",
      "    GO_POSITIVE_REGULATION_OF_T_CELL_PROLIFERATION\n",
      "    GO_DEFENSE_RESPONSE_TO_BACTERIUM\n",
      "    GO_INFLAMMATORY_RESPONSE\n",
      "65\n",
      "    5969 set.\n",
      "    ROS_GENES\n",
      "64\n",
      "    4811 set.\n",
      "    MAMMOSPHERES\n",
      "63\n",
      "    4762 set.\n",
      "    KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
      "    KEGG_PYRIMIDINE_METABOLISM\n",
      "62\n",
      "    10729 set.\n",
      "    HALLMARK_ANDROGEN_RESPONSE\n",
      "    NELSON_RESPONSE_TO_ANDROGEN_UP\n",
      "    DOANE_RESPONSE_TO_ANDROGEN_UP\n",
      "    WANG_RESPONSE_TO_ANDROGEN_UP\n",
      "61\n",
      "    6229 set.\n",
      "    REACTOME_TOLL_LIKE_RECEPTOR_CASCADES\n",
      "    NEMETH_INFLAMMATORY_RESPONSE_LPS_UP\n",
      "    KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "    SEKI_INFLAMMATORY_RESPONSE_LPS_UP\n",
      "60\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    HALLMARK_OXIDATIVE_PHOSPHORYLATION\n",
      "    HALLMARK_MYC_TARGETS_V2\n",
      "    HALLMARK_DNA_REPAIR\n",
      "6\n",
      "    4811 set.\n",
      "    GROGER_EMT (missing)\n",
      "    TAUBE_EMT (missing)\n",
      "59\n",
      "    50 set.\n",
      "    HALLMARK_E2F_TARGETS\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "58\n",
      "    4861 set.\n",
      "    REACTOME_PI3K_AKT_ACTIVATION\n",
      "    BIOCARTA_GSK3_PATHWAY\n",
      "57\n",
      "    4812 set.\n",
      "    INGA_TP53_TARGETS\n",
      "    KIM_HYPOXIA\n",
      "    LEONARD_HYPOXIA\n",
      "56\n",
      "    4762 set.\n",
      "    BAKER_HEMATOPOESIS_STAT1_TARGETS\n",
      "    DER_IFN_GAMMA_RESPONSE (missing)\n",
      "55\n",
      "    5050 set.\n",
      "    ERBB2_UP.V1_UP\n",
      "    SMID_BREAST_CANCER_ERBB2_UP\n",
      "    MEK_UP.V1_UP\n",
      "    LIM_MAMMARY_LUMINAL_MATURE_UP\n",
      "    HINATA_NFKB_TARGETS_KERATINOCYTE_UP\n",
      "    HINATA_NFKB_TARGETS_FIBROBLAST_UP\n",
      "54\n",
      "    10728 set.\n",
      "    GO_SEROTONIN_RECEPTOR_SIGNALING_PATHWAY\n",
      "    DIERICK_SEROTONIN_FUNCTION_GENES\n",
      "    REACTOME_SEROTONIN_RECEPTORS\n",
      "    GO_G_PROTEIN_COUPLED_RECEPTOR_SIGNALING_PATHWAY\n",
      "53\n",
      "    4762 set.\n",
      "    REACTOME_CELL_CYCLE\n",
      "    PID_E2F_PATHWAY\n",
      "    REACTOME_DNA_REPLICATION\n",
      "52\n",
      "    4811 set.\n",
      "    NRAS_EXTINCTION_DOWN\n",
      "    NEBEN_AML_WITH_FLT3_OR_NRAS_DN\n",
      "51\n",
      "    4762 set.\n",
      "    FEVR_CTNNB1_TARGETS_DN\n",
      "    HUPER_BREAST_BASAL_VS_LUMINAL_UP\n",
      "    BENPORATH_ES_1\n",
      "    SHEDDEN_LUNG_CANCER_POOR_SURVIVAL_A6\n",
      "    NADERI_BREAST_CANCER_PROGNOSIS_UP\n",
      "50\n",
      "    50 set.\n",
      "    HALLMARK_CHOLESTEROL_HOMEOSTASIS\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE\n",
      "5\n",
      "    4762 set.\n",
      "    PID_ALK2_PATHWAY\n",
      "    TESAR_ALK_TARGETS_EPISC_4D_UP\n",
      "    TESAR_ALK_TARGETS_EPISC_3D_UP\n",
      "49\n",
      "    6279 set.\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    HALLMARK_MYC_TARGETS_V2\n",
      "    WONG_EMBRYONIC_STEM_CELL_CORE\n",
      "48\n",
      "    5917 set.\n",
      "    GO_SKELETAL_SYSTEM_DEVELOPMENT\n",
      "47\n",
      "    189 set.\n",
      "    ALK_DN.V1_UP\n",
      "46\n",
      "    4762 set.\n",
      "    BOQUEST_STEM_CELL_UP\n",
      "    SCHUETZ_BREAST_CANCER_DUCTAL_INVASIVE_UP\n",
      "    LIM_MAMMARY_STEM_CELL_UP\n",
      "45\n",
      "    5917 set.\n",
      "    GO_PHAGOSOME_MATURATION\n",
      "    GO_PHAGOSOME_ACIDIFICATION\n",
      "44\n",
      "    6229 set.\n",
      "    REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION\n",
      "    PID_A6B1_A6B4_INTEGRIN_PATHWAY\n",
      "    NABA_CORE_MATRISOME\n",
      "    KEGG_ECM_RECEPTOR_INTERACTION\n",
      "43\n",
      "    10679 set.\n",
      "    BIOCARTA_TGFB_PATHWAY\n",
      "    GO_REGULATION_OF_STEM_CELL_PROLIFERATION\n",
      "    REACTOME_SIGNALING_BY_NODAL\n",
      "42\n",
      "    10729 set.\n",
      "    HALLMARK_ANGIOGENESIS\n",
      "    LU_TUMOR_ANGIOGENESIS_UP\n",
      "    GO_ANGIOGENESIS\n",
      "41\n",
      "    189 set.\n",
      "    EGFR_UP.V1_UP\n",
      "40\n",
      "    4762 set.\n",
      "    PUJANA_ATM_PCC_NETwork (missing)\n",
      "    PID_ATM_PATHWAY\n",
      "4\n",
      "    4762 set.\n",
      "    REACTOME_SIGNALING_BY_ACTIVATED_POINT_MUTANTS_OF_FGFR1\n",
      "    REACTOME_FGFR1_LIGAND_BINDING_AND_ACTIVATION\n",
      "39\n",
      "    4812 set.\n",
      "    NELSON_RESPONSE_TO_ANDROGEN_UP\n",
      "    HALLMARK_ANDROGEN_RESPONSE\n",
      "    DOANE_RESPONSE_TO_ANDROGEN_UP\n",
      "38\n",
      "    10679 set.\n",
      "    GO_TRANSLATION_INITIATION_FACTOR_ACTIVITY\n",
      "    GO_REGULATORY_REGION_NUCLEIC_ACID_BINDING\n",
      "    GO_TRANSLATION_REGULATOR_ACTIVITY_NUCLEIC_ACID_BINDING\n",
      "    GO_EUKARYOTIC_TRANSLATION_INITIATION_FACTOR_3_COMPLEX\n",
      "    GO_TRANSLATIONAL_INITIATION\n",
      "    REACTOME_TRANSLATION\n",
      "    GO_TRANSLATION_REGULATOR_ACTIVITY\n",
      "    GO_REGULATION_OF_TRANSLATIONAL_INITIATION\n",
      "37\n",
      "    4762 set.\n",
      "    REACTOME_REGULATION_OF_RHEB_GTPASE_ACTIVITY_BY_AMPK\n",
      "    REACTOME_REGULATION_OF_AMPK_ACTIVITY_VIA_LKB1\n",
      "    REACTOME_ENERGY_DEPENDENT_REGULATION_OF_MTOR_BY_LKB1_AMPK\n",
      "    REACTOME_ACTIVATED_AMPK_STIMULATES_FATTY_ACID_OXIDATION_IN_MUSCLE\n",
      "36\n",
      "    4762 set.\n",
      "    SEIDEN_ONCOGENESIS_BY_MET\n",
      "35\n",
      "    5001 set.\n",
      "    HALLMARK_MYC_TARGETS_V2\n",
      "    SCHLOSSER_MYC_TARGETS_AND_SERUM_RESPONSE_UP\n",
      "    SCHUHMACHER_MYC_TARGETS_UP\n",
      "    DANG_REGULATED_BY_MYC (missing)\n",
      "    MYC_UP.V1_UP\n",
      "    ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS (missing)\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS (missing)\n",
      "    PID_MYC_ACTIV_PATHWAY\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    DANG_BOUND_BY_MYC\n",
      "    SCHLOSSER_MYC_TARGETS_REPRESSED_BY_SERUM\n",
      "    BENPORATH_MYC_MAX_TARGETS\n",
      "    MYC_UP.V1 (missing)\n",
      "    COLLER_MYC_TARGETS (missing)\n",
      "    ACOSTA_PROLIFERATION_INDEPENDENT_MYC_TARGETS_UP\n",
      "34\n",
      "    4762 set.\n",
      "    BIOCARTA_CASPASE_PATHWAY\n",
      "    SA_CASPASE_CASCADE\n",
      "    PID_CASPASE_PATHWAY\n",
      "    REACTOME_NFKB_ACTIVATION_THROUGH_FADD_RIP1_PATHWAY_MEDIATED_BY_CASPASE_8_AND10\n",
      "    REACTOME_CASPASE_MEDIATED_CLEAVAGE_OF_CYTOSKELETAL_PROTEINS\n",
      "33\n",
      "    4762 set.\n",
      "    VANTVEER_BREAST_CANCER_ESR1 (missing)\n",
      "    DOANE_BREAST_CANCER_ESR1 (missing)\n",
      "    YANG_BREAST_CANCER_ESR1_UP\n",
      "32\n",
      "    4762 set.\n",
      "    REACTOME_DEGRADATION_OF_THE_EXTRACELLULAR_MATRIX\n",
      "    REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION\n",
      "    REACTOME_CELL_EXTRACELLULAR_MATRIX_INTERACTIONS\n",
      "    BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX\n",
      "    REACTOME_P130CAS_LINKAGE_TO_MAPK_SIGNALING_FOR_INTEGRINS\n",
      "31\n",
      "    4762 set.\n",
      "    PID_FOXM1_PATHWAY\n",
      "30\n",
      "    4762 set.\n",
      "    BAKKER_FOXO3_TARGETS_UP\n",
      "3\n",
      "    4951 set.\n",
      "    NFE2L2.V2\n",
      "    SINGH_NFE2L2_TARGETS\n",
      "29\n",
      "    4762 set.\n",
      "    BIOCARTA_GATA3_PATHWAY\n",
      "28\n",
      "    4762 set.\n",
      "    PID_TCR_JNK_PATHWAY\n",
      "    HAN_JNK_SINGALING_UP\n",
      "    REACTOME_JNK_C_JUN_KINASES_PHOSPHORYLATION_AND_ACTIVATION_MEDIATED_BY_ACTIVATED_HUMAN_TAK1\n",
      "27\n",
      "    4762 set.\n",
      "    REACTOME_CYCLIN_A_B1_ASSOCIATED_EVENTS_DURING_G2_M_TRANSITION\n",
      "    CHANG_CYCLING_GENES\n",
      "    REACTOME_APC_C_CDC20_MEDIATED_DEGRADATION_OF_CYCLIN_B\n",
      "268\n",
      "    6279 set.\n",
      "    KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE\n",
      "267\n",
      "    6279 set.\n",
      "    HALLMARK_MTORC1_SIGNALING\n",
      "266\n",
      "    11151 set.\n",
      "    GSE9650_EFFECTOR_VS_MEMORY_CD8_TCELL_UP\n",
      "265\n",
      "    6328 set.\n",
      "    BIDUS_METASTASIS_UP\n",
      "    CRC_proliferation_signature\n",
      "    BIDUS_METASTASIS_DN\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION\n",
      "264\n",
      "    11151 set.\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE\n",
      "    HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
      "    HALLMARK_APOPTOSIS\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_PRIMARY_INF_CD8_TCELL_UP\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_PRIMARY_INF_CD8_TCELL_DN\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_SECONDARY_INF_CD8_TCELL_UP\n",
      "    GSE30962_ACUTE_VS_CHRONIC_LCMV_SECONDARY_INF_CD8_TCELL_DN\n",
      "    GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_UP\n",
      "    GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_DN\n",
      "263\n",
      "    6229 set.\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL\n",
      "    LIM_MAMMARY_LUMINAL_PROGENITOR_UP\n",
      "    JECHLINGER_EPITHELIAL_TO_MESENCHYMAL_TRANSITION_UP\n",
      "262\n",
      "    21233 set.\n",
      "    HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION\n",
      "    GO_POSITIVE_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION\n",
      "    CORDENONSI_YAP_CONSERVED_SIGNATURE\n",
      "    SARRIO_EPITHELIAL_MESENCHYMAL_TRANSITION_UP\n",
      "261\n",
      "    6279 set.\n",
      "    HALLMARK_XENOBIOTIC_METABOLISM\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
      "    HALLMARK_COAGULATION\n",
      "260\n",
      "    6279 set.\n",
      "    HALLMARK_SPERMATOGENESIS\n",
      "    HALLMARK_MTORC1_SIGNALING\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "26\n",
      "    10679 set.\n",
      "    BOSCO_EPITHELIAL_DIFFERENTIATION_MODULE\n",
      "    GO_EPITHELIAL_CELL_DIFFERENTIATION\n",
      "259\n",
      "    11101 set.\n",
      "    GSE14308_TH17_VS_NATURAL_TREG_DN\n",
      "258\n",
      "    6229 set.\n",
      "    MOSERLE_IFNA_RESPONSE\n",
      "    REACTOME_INTERFERON_ALPHA_BETA_SIGNALING\n",
      "257\n",
      "    6279 set.\n",
      "    HALLMARK_CHOLESTEROL_HOMEOSTASIS\n",
      "    HALLMARK_INTERFERON_ALPHA_RESPONSE\n",
      "    HALLMARK_INTERFERON_GAMMA_RESPONSE\n",
      "256\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS\n",
      "    PID_E2F_PATHWAY\n",
      "    HALLMARK_G2M_CHECKPOINT\n",
      "    REACTOME_CELL_CYCLE\n",
      "    REACTOME_CELL_CYCLE_MITOTIC\n",
      "255\n",
      "    6279 set.\n",
      "    KEGG_ECM_RECEPTOR_INTERACTION\n",
      "    HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
      "254\n",
      "    6279 set.\n",
      "    HALLMARK_E2F_TARGETS\n",
      "    HALLMARK_P53_PATHWAY\n",
      "253\n",
      "    6279 set.\n",
      "    HALLMARK_MYC_TARGETS_V1\n",
      "    KIM_MYC_AMPLIFICATION_TARGETS_UP\n",
      "252\n",
      "    14765 set.\n",
      "    GO_RIBOSOME\n",
      "    GO_CYTOKINE_ACTIVITY\n",
      "    GO_MATURE_RIBOSOME_ASSEMBLY\n",
      "    GO_CYTOSOLIC_RIBOSOME\n",
      "        Is KS < extreme: NaN\n",
      "        Is KS < area: NaN\n",
      "        Is KS <> extreme: NaN\n",
      "        Is KS <> area: NaN\n",
      "        Is JSD2 < extreme: NaN\n",
      "        Is JSD2 < area: NaN\n",
      "        Is JSD2 <> extreme: NaN\n",
      "        Is JSD2 <> area: NaN\n",
      "        A KS < extreme: NaN\n",
      "        A KS < area: NaN\n",
      "        A KS <> extreme: NaN\n",
      "        A KS <> area: NaN\n",
      "        A JSD2 < extreme: NaN\n",
      "        A JSD2 < area: NaN\n",
      "        A JSD2 <> extreme: NaN\n",
      "        A JSD2 <> area: NaN\n",
      "        A JSD2w < extreme: NaN\n",
      "        A JSD2w < area: NaN\n",
      "        A JSD2w <> extreme: NaN\n",
      "        A JSD2w <> area: NaN\n",
      "        Bad gene set.\n",
      "    GO_ENDOTHELIAL_CELL_MATRIX_ADHESION\n",
      "    GO_REGULATION_OF_CHAPERONE_MEDIATED_PROTEIN_FOLDING\n",
      "    GO_PROTEIN_FOLDING_IN_ENDOPLASMIC_RETICULUM\n",
      "251\n",
      "    6278 set.\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL\n",
      "    VERHAAK_GLIOBLASTOMA_PRONEURAL\n",
      "    PHILLIPS_GLIOBLASTOMA_MESENCHYMAL\n",
      "    PHILLIPS__GLIOBLASTOMA_PRONEURAL\n",
      "    ANASTASSIOU_MULTICANCER_INVASIVENESS\n",
      "    SEGERMAN_MULTITHERAPY_RESISTANCE\n",
      "250\n",
      "    5529 set.\n",
      "    BROWNE_INTERFERON_RESPONSIVE_GENES\n",
      "    REACTOME_CHOLESTEROL_BIOSYNTHESIS\n",
      "25\n",
      "    4762 set.\n",
      "    REACTOME_ACTIVATED_TAK1_MEDIATES_P38_MAPK_ACTIVATION\n",
      "    BIOCARTA_P38MAPK_PATHWAY\n",
      "249\n",
      "    5529 set.\n",
      "    HSIAO_HOUSEKEEPING_GENES\n",
      "    WONG_EMBRYONIC_STEM_CELL_CORE\n",
      "    REACTOME_TRANSLATION\n",
      "    RHEIN_ALL_GLUCOCORTICOID_THERAPY_DN\n",
      "    REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE\n",
      "    SPIELMAN_LYMPHOBLAST_EUROPEAN_VS_ASIAN_UP\n",
      "248\n",
      "    5529 set.\n",
      "    REACTOME_GLUCOSE_METABOLISM\n",
      "    REACTOME_GLYCOLYSIS\n",
      "    KEGG_PYRUVATE_METABOLISM\n",
      "    KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "247\n",
      "    4762 set.\n",
      "    YU_MYC_TARGETS_UP\n",
      "246\n",
      "    4762 set.\n",
      "    VERHAAK_GLIOBLASTOMA_PRONEURAL\n",
      "    VERHAAK_GLIOBLASTOMA_MESENCHYMAL\n",
      "    VERHAAK_GLIOBLASTOMA_CLASSICAL\n",
      "    PID_WNT_CANONICAL_PATHWAY\n",
      "    PID_NOTCH_PATHWAY\n",
      "245\n",
      "    4951 set.\n",
      "    WNT_UP.V1_UP\n",
      "    KEGG_MAPK_SIGNALING_PATHWAY\n",
      "    VERHAAK_AML_WITH_NPM1_MUTATED_DN\n",
      "    BIOCARTA_CELLCYCLE_PATHWAY\n",
      "    MYC_UP.V1_UP\n",
      "244\n",
      "    10401 set.\n",
      "    GSE25085_FETAL_LIVER_VS_ADULT_BM_SP4_THYMIC_IMPLANT_DN\n",
      "243\n",
      "    5529 set.\n",
      "    REACTOME_FGFRL1_MODULATION_OF_FGFR1_SIGNALING\n",
      "242\n",
      "    5529 set.\n",
      "    REACTOME_SEMA3A_PLEXIN_REPULSION_SIGNALING_BY_INHIBITING_INTEGRIN_ADHESION\n",
      "241\n",
      "    5529 set.\n",
      "    REACTOME_DEFECTIVE_CHSY1_CAUSES_TPBS\n",
      "240\n",
      "    6829 set.\n",
      "    IPA_TCF12_UP\n",
      "24\n",
      "    10679 set.\n",
      "    "
     ]
    }
   ],
   "source": [
    "for benchmark_directory_path in reverse(read_directory(benchmarks_directory_path))\n",
    "\n",
    "    #\n",
    "    name = splitpath(benchmark_directory_path)[end]\n",
    "\n",
    "    println(name)\n",
    "\n",
    "    #\n",
    "    element_, score_ =\n",
    "        eachcol(read_data(joinpath(benchmark_directory_path, \"gene_x_score.tsv\")))\n",
    "\n",
    "    json_dict = JSON.parse(open(joinpath(benchmark_directory_path, \"gene_sets.json\")))\n",
    "\n",
    "    #\n",
    "    set_to_element_ = Dict{String, Vector{String}}()\n",
    "\n",
    "    for gmt_path in json_dict[\"gene_sets_tested\"]\n",
    "\n",
    "        merge!(set_to_element_, name_to_set_to_element_[splitpath(gmt_path)[end]])\n",
    "\n",
    "    end\n",
    "\n",
    "    #\n",
    "    set_ = sort(collect(keys(set_to_element_)))\n",
    "\n",
    "    n_set = length(set_)\n",
    "\n",
    "    println(\"    \", n_set, \" set.\")\n",
    "\n",
    "    #\n",
    "    set_to_method_to_result = score_set_new(element_, score_, set_to_element_)\n",
    "\n",
    "    #\n",
    "    set_x_method = Matrix{Float64}(undef, n_set, n_method)\n",
    "\n",
    "    for (set_i, (set, method_to_result)) in enumerate(set_to_method_to_result)\n",
    "\n",
    "        set_x_method_row = []\n",
    "\n",
    "        for result in values(method_to_result)\n",
    "\n",
    "            append!(set_x_method_row, result[2:3])\n",
    "\n",
    "        end\n",
    "\n",
    "        set_x_method[set_i, :] = set_x_method_row\n",
    "\n",
    "    end\n",
    "\n",
    "    #\n",
    "    for set in json_dict[\"gene_sets_positive\"]\n",
    "\n",
    "        #\n",
    "        if in(set, set_)\n",
    "\n",
    "            println(\"    \", set)\n",
    "\n",
    "        else\n",
    "\n",
    "            println(\"    \", set, \" (missing)\")\n",
    "\n",
    "            continue\n",
    "\n",
    "        end\n",
    "\n",
    "        #\n",
    "        benchmark_x_method_row = Vector{Float64}()\n",
    "\n",
    "        #\n",
    "        for (method_i, set_score_) in enumerate(eachcol(set_x_method))\n",
    "\n",
    "            #\n",
    "            is_negative_ = set_score_ .< 0\n",
    "\n",
    "            is_positive_ = 0 .<= set_score_\n",
    "\n",
    "            #\n",
    "            set_negative_ = set_[is_negative_]\n",
    "\n",
    "            set_positive_ = set_[is_positive_]\n",
    "\n",
    "            #\n",
    "            negative_ = .-normalize(.-set_score_[is_negative_], \"1224\") / sum(is_negative_)\n",
    "\n",
    "            positive_ = normalize(set_score_[is_positive_], \"1224\") / sum(is_positive_)\n",
    "\n",
    "            #\n",
    "            i_negative = findfirst(set_negative_ .== set)\n",
    "\n",
    "            i_positive = findfirst(set_positive_ .== set)\n",
    "\n",
    "            if i_negative == nothing && i_positive == nothing\n",
    "\n",
    "                println(\"        \", method_[method_i], \": \", set_score_[findfirst(set_ .== set)])\n",
    "\n",
    "                continue\n",
    "\n",
    "            elseif i_negative != nothing\n",
    "\n",
    "                benchmark_score = negative_[i_negative]\n",
    "\n",
    "            elseif i_positive != nothing\n",
    "\n",
    "                benchmark_score = positive_[i_positive]\n",
    "\n",
    "            end\n",
    "\n",
    "            push!(benchmark_x_method_row, benchmark_score)\n",
    "\n",
    "        end\n",
    "\n",
    "        #\n",
    "        if 0 == length(benchmark_x_method_row)\n",
    "\n",
    "            println(\"        Bad gene set.\")\n",
    "\n",
    "        else\n",
    "\n",
    "            push!(benchmark_, string(name, '.', set))\n",
    "\n",
    "            push!(benchmark_x_method_rows, benchmark_x_method_row)\n",
    "\n",
    "        end\n",
    "\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "459b6a2c-864b-4e0c-873e-0b6cc1ce71e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_x_method = convert_vector_of_vector_to_matrix(benchmark_x_method_rows)\n",
    "\n",
    "benchmark_x_method = DataFrame(benchmark_x_method, method_)\n",
    "\n",
    "insertcols!(benchmark_x_method, 1, Symbol(\"Benchmark\") => benchmark_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6d436a0-f4c7-4ec5-9e96-3f2693ecefce",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_x_method_path = \"benchmark_x_method.tsv\"\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dbed1ad-ca70-4067-905f-79392774757c",
   "metadata": {},
   "outputs": [],
   "source": [
    "CSV.write(benchmark_x_method_path, benchmark_x_method)\n",
    "\n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303533d3-59ae-46eb-b055-d1e31b91ff1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_x_method = read_data(benchmark_x_method_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbb84c7e-19df-4d5a-aa1a-0d49bb3cd1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_x_method_is = Matrix{}(undef, length(benchmark_), n_method)\n",
    "\n",
    "for (benchmark_i, benchmark_score_) in enumerate(eachrow(select(benchmark_x_method, Not(Symbol(\"Benchmark\")))))\n",
    "    \n",
    "    classic_score = benchmark_score_[Symbol(\"Is KS < extreme\")]\n",
    "    \n",
    "    benchmark_score_ = Vector(benchmark_score_)\n",
    "    \n",
    "    if classic_score < 0\n",
    "        \n",
    "        is_ = benchmark_score_ .< classic_score\n",
    "        \n",
    "    else\n",
    "        \n",
    "        is_ = classic_score .< benchmark_score_\n",
    "        \n",
    "    end\n",
    "    \n",
    "    benchmark_x_method_is[benchmark_i, :] = is_\n",
    "    \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef02e1f0-1e67-45a4-8759-bb048d327bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (method, n_better) in zip(sort_like(sum(eachrow(benchmark_x_method_is)), method_)...)\n",
    "    \n",
    "    println(method, \"    \", n_better)\n",
    "    \n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.1",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
